---
title: "lcWGS"
author: "Laura Spencer"
date: "2023-08-29"
output: html_document
---

```{r, message=FALSE, warning=FALSE, results=FALSE}
#### Load libraries and source scripts 

`%!in%` = Negate(`%in%`)
source("../scripts/biostats.R")


# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("tidyverse", "readxl", "janitor", "purrr", "ggpubr")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c()

# # Install BiocManager if needed
# if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# 
# # Get names of all required packages that aren't installed
# new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
# new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# # Install all new packages
# if(length(new.packages)) install.packages(new.packages)
# if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})
```
From Ingrid: Shumagins, WKodiak, PWS, and Cook Inlet

Sara's presentation: https://docs.google.com/presentation/d/1t11nqJNUMMkaulUN4HrPFqEGznRV0Fi84BwayesOUXA/edit?usp=sharing

Amchitka
Adak
Kodiak
West Kodiak
Shumagins
Unimak
Hecate Strait
Pervenets
How many samples per? 
Fish might move from eastern to western gulf 

Prepare list of file names for 10 samples from each of the populations of interest: 

```{r}
ref.10 <- read_delim("../references/ABLGs_10inds_per_pop.txt", delim = "/t", col_names = "ID")

ref.10.files <- read_delim("../references/pcod_lcwgs_reference.txt", delim = "/t", col_names = "file") %>% 
  separate(file, into = c("ID", "rest"), remove = FALSE) %>% select(-rest) %>%
  mutate(ID=as.numeric(gsub("ABLG", "", ID))) %>%
  filter(ID %in% ref.10$ID) %>% select(file) %>% write_delim(file = "../references/pcod_10refs.txt", col_names = F)
```


Read in sample metadata

```{r, warning=FALSE}
sample.info <- read_excel("../../Pcod Temp Growth experiment 2022-23 DATA.xlsx", sheet = "AllData") %>% clean_names() %>%
  mutate_at(c("tank", "temperature"), factor) %>%
  select(temperature, genetic_sampling_count) %>% 
  rename(sample=genetic_sampling_count, treatment=temperature) %>%
  rbind(
    read_excel("../references/20230414_pcod_named.xlsx") %>% clean_names() %>%
      select(ablg, location1) %>% rename(sample=ablg, treatment=location1)) 
```

Read in depth coverage results from alignment to look for bad samples 

```{r}
depths <- 
#  read_delim(file = "../analysis-20230823/pcod-lcWGS_depths.csv", 
  read_delim(file = "../analysis-20230922/pcod-lcWGS_depths.csv", 
             delim = "\t",  col_names = c("sample", "ave_depth")) %>%
  mutate(sample=gsub("GM|ABLG", "", sample)) %>%
  mutate(sample=as.numeric(sample))

depths %>%
  left_join(sample.info) %>%
  ggplot(aes(x=reorder(sample, ave_depth), y=ave_depth, fill=treatment)) + 
  geom_bar(stat="identity") + 
  theme_minimal() +
  theme(axis.text.x = element_text(angle=90, vjust=0.5, size=8.5)) #+
#  scale_fill_manual(values = rev(c("#d7191c", "#fdae61", "#abd9e9", "#2c7bb6")))
```

Perform PCA on each covariate matrix resulting from the lcWGS pipeline. One PCA for each chromosome + one whole genome 

```{bash}
#rm ../analysis-20230823/pca/cov-results.txt #rm file if it already exists
rm ../analysis-20230922/pca/cov-results.txt #rm file if it already exists
#for file in ../analysis-20230823/pca/*.cov
for file in ../analysis-20230922/pca/*.cov
do
filename="$(echo $file)"
sample="$(basename -a $filename | cut -d "." -f 1)"
#printf "%s\t%s\n" "$filename" "$sample" >> ../analysis-20230823/pca/cov-results.txt
printf "%s\t%s\n" "$filename" "$sample" >> ../analysis-20230922/pca/cov-results.txt
done
```

```{r}
outlier <- c("GM121") #c("GM121", "ABLG2109", "ABLG2937", "ABLG2622") NULL #

sample.order <- (read_delim(
#  file="../analysis-20230823/pca/pcod-lcWGS_filtered_bamslist.txt", 
  file="../analysis-20230922/pcod-lcWGS_filtered_bamslist.txt", 
                            delim = "/t", col_names = "sample") %>%
                   mutate(
#                     sample=gsub("/home/lspencer/pcod-lcwgs-2023/analysis-20230828/bamtools/pcod-lcWGS_|_sorted_dedup_clipped.bam", 
                     sample=gsub("/home/lspencer/pcod-lcwgs-2023/analysis-20230922/bamtools/pcod-lcWGS_|_sorted_dedup_clipped.bam", 
                                      "", sample)))$sample
                   

filenames <- read_delim(
#  file="../analysis-20230823/pca/cov-results.txt", 
  file="../analysis-20230922/pca/cov-results.txt", 
  col_names = c("filename", "chromosome"), delim = "\t") %>%
  mutate(chromosome=gsub("pcod-lcWGS_|_polymorphic|-polymorphic", "", chromosome))
files <- file.path(filenames$filename) #extract vector of filenames
all(file.exists(files)) #easy code to check that all files exist!

chrom_covar <- vector(mode = "list", length = nrow(filenames))
names(chrom_covar) <- c(filenames$chromosome)

for (i in 1:nrow(filenames)) {
    chrom_covar[[i]] <- read_delim(file=files[i], col_names = sample.order) %>% 
      dplyr::select(-outlier) %>% as.matrix()
}

#(covariances.chrom <- chrom_covar %>% purrr::reduce(full_join, by = "gene") %>% column_to_rownames(var="gene"))
```

### PCAs using prcomp

This enables screeplot to ID significant PCs, calculates variance, etc. 


Separate PCA for each chromosome (#1-23) and for the whole genome

```{r}
# Filter sample info to only include those we used in our lcWGS pipeline 
sample.info <- sample.info %>% 
  mutate(sample.id=case_when(
      (treatment=="0" | treatment=="5" | treatment=="9" | treatment=="16") ~ paste("GM", sample, sep=""), 
      TRUE ~ paste("ABLG", sample, sep=""))) %>%
  filter(sample.id %in% sample.order)

# Get sample IDs for those from our likely spawning populations, and experimental animals (i.e. those to include in a PCA)
likely.pops <- (sample.info %>%
                  filter(sample.id != outlier) %>% #remove outlier sample(s)
        filter(treatment %in% c("0", "5", "9", "16", "Kodiak", "Shumagins", "HecateStrait")))$sample.id #keep only likely source pops
```

```{r, warning=F, message=F}
pcas<-list(1:length(filenames$chromosome))
for (i in 1:length(filenames$chromosome)){

    print(names(chrom_covar[i]))
  covar<- chrom_covar[[i]]
  pca.princomp <- prcomp(covar[,likely.pops], scale=F) #scale=F for variance-covariance matrix
  #pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
  pc.percent <- pca.eigenval(pca.princomp)[2,1:6]*100
  screeplot(pca.princomp, bstick=FALSE) 
  pc.percent[1:2] %>% sum()
  
  #### Generate dataframe with prcomp results 
  tab.expr <- data.frame(sample.id = colnames(covar[,likely.pops]),
      EV1.all = pca.princomp$rotation[,1],    # the first eigenvector
      EV2.all = pca.princomp$rotation[,2],    # the second eigenvector
      EV3.all = pca.princomp$rotation[,3],    # the third eigenvector
      EV4.all = pca.princomp$rotation[,4],    # the fourth eigenvector
      EV5.all = pca.princomp$rotation[,5],    # the fourth eigenvector
      EV6.all = pca.princomp$rotation[,6],    # the fourth eigenvector
      stringsAsFactors = FALSE)
  #shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
  #hist(pca.princomp$x) #normal? hard to say, maybe
  tab.expr.annot <- left_join(tab.expr %>% mutate(sample.id=as.numeric(sub("GM|ABLG", "", sample.id))), 
                              sample.info[c("treatment", "sample")], by=c("sample.id"="sample")) %>% droplevels() %>%
    mutate(treatment=factor(treatment, ordered = T, 
                            levels = c("0", "5", "9", "16", 
                                       "Russia", "Pervenets", "TanagaIsland", 
                                       "AmchitkaPass", "Unimak", "Shumagins", 
                                       "Kodiak", "HecateStrait")))
  
  
  pcas[[i]] <- print(ggscatter(tab.expr.annot,
            group=c("temperature"),
            x="EV1.all", y="EV2.all", col="treatment", size=3, alpha=0.85, 
            ellipse = FALSE, star.plot = T) + #, label="sample.id"
    theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
    xlab(paste("PC1 (", round(pc.percent[1], digits = 2), "%)", sep="")) + 
    ylab(paste("PC2 (", round(pc.percent[2], digits = 2), "%)", sep="")) + 
    theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
    # scale_color_manual(name="Temperature/Population", 
    #                    values=c("0"="purple", "5"="blue", "9"="orange", "16"="red")) +
    scale_color_manual(name="Temperature/Population",
                       values=c("0"="gray90", "5"="gray50", "9"="gray25", "16"="black",
                                "Russia"="navyblue", "Pervenets"="cornflowerblue",
                                "TanagaIsland"="#31a354", "AmchitkaPass"="#74c476",
                                "Unimak"="antiquewhite", "Shumagins"="mistyrose",
                                "Kodiak"="palevioletred2", "HecateStrait"="purple")) +
      ggtitle(names(chrom_covar[i])))
}

# pdf('../analysis-20230922/pca/pcas12-by-chrom.pdf')
# pcas
# dev.off()
# 
# # Save objects for Plotly in RStudio Cloud 
# save(filenames, file = "../analysis-20230922/pca/plotly/filenames")
# save(chrom_covar, file = "../analysis-20230922/pca/plotly/chrom_covar")
# save(sample.info, file = "../analysis-20230922/pca/plotly/sample.info")
# save(pcas, file="../analysis-20230922/pca/plotly/pcas")
```

Whole-Genome PCAs using more PC axes 

```{r}
pca.princomp <- prcomp(chrom_covar$wholegenome[,likely.pops], scale=F) #scale=F for variance-covariance matrix
  #pca.eigenval(pca.princomp) #The Proporation of Variance = %variance 
  pc.percent <- pca.eigenval(pca.princomp)[2,1:6]*100
  screeplot(pca.princomp, bstick=FALSE) 
  pc.percent[1:6] %>% sum()
  
  #### Generate dataframe with prcomp results 
  tab.expr <- data.frame(sample.id = colnames(chrom_covar$wholegenome[,likely.pops]),
      EV1.all = pca.princomp$rotation[,1],    # the first eigenvector
      EV2.all = pca.princomp$rotation[,2],    # the second eigenvector
      EV3.all = pca.princomp$rotation[,3],    # the third eigenvector
      EV4.all = pca.princomp$rotation[,4],    # the fourth eigenvector
      EV5.all = pca.princomp$rotation[,5],    # the fourth eigenvector
      EV6.all = pca.princomp$rotation[,6],    # the fourth eigenvector
      stringsAsFactors = FALSE)
  #shapiro.test(pca.princomp$x) #sample size too large for shapiro test which is weird 
  #hist(pca.princomp$x) #normal? hard to say, maybe
  tab.expr.annot <- left_join(tab.expr %>% mutate(sample.id=as.numeric(sub("GM|ABLG", "", sample.id))), sample.info[c("treatment", "sample")], by=c("sample.id"="sample")) %>% droplevels() %>%
    mutate(treatment=factor(treatment, ordered = T, 
                            levels = c("0", "5", "9", "16", 
                                       "Russia", "Pervenets", "TanagaIsland", 
                                       "AmchitkaPass", "Unimak", "Shumagins", 
                                       "Kodiak", "HecateStrait")))

axes <- data.frame("pc.x"=c(1,1,1,1,1,2,2,2,2,3,3,3,4,4,5), 
                   "pc.y"=c(2,3,4,5,6,3,4,5,6,4,5,6,5,6,6)) %>%
  mutate(ev.x=paste("EV", pc.x, ".all", sep=""), ev.y=paste("EV", pc.y, ".all", sep=""))
variance <- pc.percent %>% as.data.frame() %>% set_names("variance") %>% rownames_to_column("axis") %>% 
  mutate(axis=as.numeric(gsub("PC", "", axis))) 

pcas.genome<-list(1:nrow(axes))
for (i in 1:nrow(axes)){
  pcas.genome[[i]] <- print(ggscatter(tab.expr.annot,
            group=c("temperature"),
            x=axes[i,"ev.y"], y=axes[i,"ev.x"], col="treatment", size=3, alpha=0.85, 
            ellipse = FALSE, star.plot = T) + #, label="sample.id"
    theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
    xlab(paste("PC", axes[i, "pc.y"], " (", round(variance[variance$axis==axes[i, "pc.y"], "variance"], digits = 2), "%)", sep="")) + 
    ylab(paste("PC", axes[i, "pc.x"], " (", round(variance[variance$axis==axes[i, "pc.x"], "variance"], digits = 2), "%)", sep="")) + 
    theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
    scale_color_manual(name="Temperature/Population", 
                       values=c("0"="gray75", "5"="gray50", "9"="gray25", "16"="black",
                                "Russia"="navyblue", "Pervenets"="cornflowerblue",
                                "TanagaIsland"="#31a354", "AmchitkaPass"="#74c476", 
                                "Unimak"="antiquewhite", "Shumagins"="mistyrose", 
                                "Kodiak"="palevioletred2", "HecateStrait"="purple")) +
      ggtitle(paste("PC", axes[i, "pc.y"], "x", "PC", axes[i, "pc.x"], sep="")))
}

pdf('../analysis-20230922/pca/pcas-genomewide.pdf', width = 8.5, height = 7)
pcas.genome
dev.off()

```

Zoom in on temperature experiment animals in PC1xPC2 space to look at slight separation of 16C animals - why would that be? 

```{r}
i<-1
ggscatter(tab.expr.annot,
            group=c("temperature"),
            x=axes[i,"ev.x"], y=axes[i,"ev.y"], col="treatment", size=3, alpha=0.85, 
            ellipse = T, star.plot = F) + #, label="sample.id"
    theme_minimal() + ggtitle("Global gene expression PC1xPC2") + 
    xlab(paste("PC", axes[i, "pc.x"], " (", round(variance[variance$axis==axes[i, "pc.x"], "variance"], digits = 2), "%)", sep="")) + 
    ylab(paste("PC", axes[i, "pc.y"], " (", round(variance[variance$axis==axes[i, "pc.y"], "variance"], digits = 2), "%)", sep="")) + 
    theme(legend.position = "right", legend.text=element_text(size=8), legend.title=element_text(size=9)) + 
    scale_color_manual(name="Temperature/Population", 
                       values=c("0"="#2c7bb6", "5"="#abd9e9", "9"="#fdae61", "16"="#d7191c")) +
    scale_fill_manual(name="Temperature/Population", 
                       values=c("0"=lighten("#2c7bb6"), "5"=lighten("#abd9e9"), "9"=lighten("#fdae61"), "16"=lighten("#d7191c"))) +
      ggtitle(paste("Experimental animals only, PC", axes[i, "pc.x"], "x", "PC", axes[i, "pc.y"], sep="")) + 
  xlim(-.1,-.01) + ylim(-0.12, 0.1)
```


Genome wide PCA using Sara's code 

```{r}
pca <- as.matrix(read.table("../analysis-20230922/pca/pcod-lcWGS_wholegenome-polymorphic.cov"))
pca_e <- eigen(pca)
prop_var <-  round(pca_e$values/sum(pca_e$values)*100,2)
first3pcs <- data.frame(pca_e$vectors[,1:3])
dim(pca)
pca_df <- cbind(as.numeric(gsub("GM|ABLG", "", sample.order)), first3pcs) %>% 
  set_names(c("sample", "PC1", "PC2", "PC3")) %>%
  left_join(sample.info) %>%
    mutate(treatment=factor(treatment, ordered = T, 
                          levels = c("0", "5", "9", "16", 
                                     "Russia", "Pervenets", "TanagaIsland", 
                                     "AmchitkaPass", "Unimak", "Shumagins", 
                                     "Kodiak", "HecateStrait")))


(genomewidePlot <- ggplot(data = pca_df %>% filter(sample!=121), 
                          aes(x = PC1, y = PC2, color = treatment)) + 
  geom_point(aes(x = PC1 , y = PC2, color = treatment), size = 3) + 
  ggtitle(paste0("Pacific cod whole genome PCA\nExperimental fish and reference populations")) +
    scale_color_manual(name="Temperature/Population", 
                       values=c("0"="black", "5"="gray25", "9"="gray50", "16"="gray90",
                                "Russia"="navyblue", "Pervenets"="cornflowerblue",
                                "TanagaIsland"="#31a354", "AmchitkaPass"="#74c476", 
                                "Unimak"="antiquewhite", "Shumagins"="mistyrose", 
                                "Kodiak"="palevioletred2", "HecateStrait"="purple")) +
  #scale_shape_manual(values = myshapes) +
  xlab(paste0("PC1 - ", format(round(prop_var[1], 2), nsmall = 2), "%")) +
  ylab(paste0("PC2 - ", format(round(prop_var[2], 2), nsmall = 2), "%")) +
  theme_bw() + 
  theme(panel.border = element_blank(), panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")))

pdf('../analysis-20230922/pca/pcas-genomewide-sara-code.pdf', width = 8.5, height = 7)
genomewidePlot
dev.off()
save(genomewidePlot, file="../analysis-20230922/pca/plotly/genomewidePlot")
```


Look at size distribution of P. cod by temperature 

```{r}
phenotypes <- read_excel("../../Pcod Temp Growth experiment 2022-23 DATA.xlsx", sheet = "AllData") %>% clean_names() %>%
  mutate_at(c("tank", "temperature", "microchip_id", "dissection_date", "genetic_sampling_count"), factor) %>%
  rename("sl_final"="sl_mm", "wwt_final"="whole_body_ww_g") #%>% 
  # pivot_longer(cols = c("sl_11212022", "wwt_11212022", 
  #                       "sl_12272022", "wwt_12272022", 
  #                       "sl_final", "ww_final",
  #                       "total_liver_ww_mg", "liverfor_lipids_ww_mg", "muscle_w_wfor_lipids_mg"),names_to = "metric", values_to = "value")

# Liver wet weight by temperature 
phenotypes %>%
  ggplot() + geom_boxplot(aes(x=temperature, y=total_liver_ww_mg, fill=temperature)) + theme_minimal()
summary(aov(total_liver_ww_mg~temperature, phenotypes))
TukeyHSD(aov(total_liver_ww_mg~temperature, phenotypes)) #16C smaller liver than other temperatures

# Summarize by temperature - mean and SD
phenotypes %>%
  group_by(temperature) %>%
  summarise(sl_mean.1=mean(sl_11212022),sl_sd.1=sd(sl_11212022), 
            sl_mean.2=mean(sl_12272022),sl_sd.2=sd(sl_12272022),
            sl_mean.3=mean(sl_final),sl_sd.3=sd(sl_final),
            
            wwt_mean.1=mean(wwt_11212022),wwt_sd.1=sd(wwt_11212022), 
            wwt_mean.2=mean(wwt_12272022),wwt_sd.2=sd(wwt_12272022),
            wwt_mean.3=mean(wwt_final),wwt_sd.3=sd(wwt_final)) %>%
  pivot_longer(cols = -temperature) %>%
  separate(name, sep = "\\.", into = c("metric", "time")) %>%
  mutate_at(c("metric"), factor) %>% mutate(time=as.numeric(time)) %>%
  filter(metric=="wwt_mean") %>%
  ggplot() + geom_line(aes(x=time, y=value, color=temperature)) + theme_minimal() +
#  ggtitle("Length")
  ggtitle("Weight")
```

NEW PCA
 - SHUMAGINS (VERY SIMILAR TO KODIAK) 
 - KODIAK 
 - HECATE STRAIT 
 
```{r}
tab.expr.annot
```
 

