---
title: "05-DEXSeq-exons"
author: "Laura Spencer"
date: "2024-05-02"
output: html_document
---

```{r, message=FALSE, warning=FALSE, results=FALSE}
#### Load libraries and source scripts 

source("../../references/load_SubreadOutput.R")
`%!in%` = Negate(`%in%`)

# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("tidyverse", "plotly", "janitor", "eulerr")

# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("DEXSeq", "BiocParallel", "goseq")

# # Install BiocManager if needed
# if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
# 
# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})
```

Here I use [DEXSeq](https://bioconductor.org/packages/release/bioc/vignettes/DEXSeq/inst/doc/DEXSeq.html) with exon counts to examine differential exon usage in response to temperature, which suggests alternative splicing for possibly optimizing protein performance.  

YAY! - "DEXSeq is not limited to simple two-group comparisons; rather, it uses so-called generalized linear models (GLMs) to permit ANOVA-like analysis of potentially complex experimental designs."

As per the DEXSeq vignette, Vivek Bhardwaj provides a function to import the count files into R and generate a DEXSeqDataSet object in his github: https://github.com/vivekbhr/Subread_to_DEXSeq/tree/master, which I will use here

### Load output from featureCounts into DEXSeq

Prepare sampleData data.frame, which contains sample names used for featurecounts as rownames, plus condition, and other variables you want to use for DEXSeq design matrix.

Example :

samp <- data.frame(row.names = c("cont_1","cont_2","test_1","test_2"), 
                        condition = rep(c("control","trt"),each=2))
dxd.fc <- DEXSeqDataSetFromFeatureCounts("dm6_fCount.out",
                                         flattenedfile = "dm6_ens76_flat.gtf",sampleData = samp)
This will create a dxd object that you can use for DEXSeq analysis.

```{bash}
head "../aligned/featurecounts_exon_dexseq"
```

Remove extraneous text from sample names 

```{bash}
sed -e 's#/home/lspencer/pcod-juv-temp/aligned/#sample_#g' ../aligned/featurecounts_exon_dexseq > ../aligned/../aligned/featurecounts_exon_dexseq.renamed
sed -i -e 's#.Aligned.sortedByCoord.out.bam##g'  ../aligned/featurecounts_exon_dexseq.renamed
```


```{bash}
head -n 2 ../aligned/featurecounts_exon_dexseq.renamed
```
### Prefilter featureCounts matrix to remove lowly expressed exons, remove samples, etc. 

### Import counts generated using featurecounts from star alignments

```{r}
exon.counts.fc <- data.frame(read.table("../aligned/featurecounts_exon_dexseq.renamed", header = T, stringsAsFactors = F, fill = FALSE)) 
```

### Summarize counts and visualize

```{r}
print(paste("Number of samples:", ncol(exon.counts.fc %>% dplyr::select(contains("sample"))), sep=" "))
print(paste("Total number of exons in dataframe:", prettyNum(nrow(exon.counts.fc), big.mark = ","), sep=" "))
print(paste("Average number of exons per sample:", prettyNum(mean(colSums(exon.counts.fc %>% dplyr::select(contains("sample")) != 0)) %>% round(), big.mark = ","), sep=" "))
print(paste("Total counts, all samples:", prettyNum(sum(colSums(exon.counts.fc %>% dplyr::select(contains("sample")))), big.mark = ","), sep=" "))

#inspect total counts by sample 
# sample 149 very high 
# sample 129 kinda low
ggplotly(
   ggplot(data.frame(colSums(exon.counts.fc %>% dplyr::select(contains("sample")))) %>% 
            dplyr::rename(count.total = 1) %>% rownames_to_column(var="sample")) + 
     geom_bar(aes(x=sample, y=count.total), stat = "identity") + ggtitle("Total count by sample") + 
              theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())) 
```

## OPTIONAL: FILTER WHOLE SAMPLES. 

Remove whole samples from the data set and sample info here if needed

```{r}
outliers <- c("sample_31", "sample_41", "sample_149", "sample_129")
remove.list <- c(exon.counts.fc %>% select(contains(c(".S", ".G"))) %>% colnames(), outliers) #remove list for featureCounts-generated counts
exon.counts.fc <- exon.counts.fc[ , which(names(exon.counts.fc) %!in% remove.list)]
```

### Add arbritary rownames 

```{r}
rownames(exon.counts.fc) <- paste("exon", 1:nrow(exon.counts.fc), sep=".")
rownames(exon.counts.fc[-1:-6]) %>% head(n=25)
exon.counts.fc.t <- t(exon.counts.fc[-1:-6])
```


### Pre-filtering - remove low-frequency exons 

```{r}
keep1 <- rowMeans(exon.counts.fc[-1:-6], na.rm=TRUE) >= 10 #identify exons with mean count >= 10 across all samples (excluding NAs = 10)
keep2 <- rowSums( exon.counts.fc >= 10 ) >= 0.1*69 #identify exons with counts>=10 across at minimum 10% of the samples
keep <- unique(c(names(which(keep1 == T)), names(which(keep2 == T)))) # list of exons meeting one of the two above criteria
exon.counts.fc.s <- exon.counts.fc[keep,]

print(paste("# exons before filtering:", nrow(exon.counts.fc)))
print(paste("# exons remaining after pre-filtering:", nrow(exon.counts.fc.s)))
print(paste("# of exons dropped:", nrow(exon.counts.fc) - nrow(exon.counts.fc.s), sep=" "))
print(paste("% of fragments remaining after pre-filtering: ", signif(100*sum(exon.counts.fc.s[-1:-6])/sum(exon.counts.fc[-1:-6]), digits = 5), "%", sep=""))
print(paste("Number of fragments dropped: ", signif(sum(exon.counts.fc[-1:-6])-sum(exon.counts.fc.s[-1:-6]), digits = 5)))
print(paste("% of fragments dropped: ", signif(100*(sum(exon.counts.fc[-1:-6])-sum(exon.counts.fc.s[-1:-6]))/sum(exon.counts.fc[-1:-6]), digits = 5), "%", sep=""))
print(paste("Number of fragments remaining: ", signif(sum(exon.counts.fc.s[-1:-6]), digits = 5)))
```
```{r}
View(exon.counts.fc.s)
```


```{r}
write_delim(exon.counts.fc.s, file = "../aligned/featurecounts_exon_dexseq_filtered", delim = "\t", col_names = T)
```

### Now read the filtered exon count matrix into R as a DEXSeq object

```{r}
b <- c(data.frame(read.table("../aligned/featurecounts_exon_dexseq_filtered", header = T, stringsAsFactors = F, fill = FALSE)) %>% 
  dplyr::select(starts_with("sample_")) %>% colnames())

samp <- b %>% as.data.frame() %>% 
  set_names("sample_name") %>% mutate(sample_number=gsub("sample_", "", sample_name)) %>%
  left_join(
    read_delim("../../data/DESeq2_Sample_Information.txt", delim="\t") %>% 
      clean_names() %>% 
      mutate(condition=case_when(
        temp_treatment=="9" ~ "a_9",
        temp_treatment=="0" ~ "b_0",
        temp_treatment=="5" ~ "c_5",
        temp_treatment=="16" ~ "d_16")) %>%
      mutate(condition=factor(as.character(condition)),
             sample_number=as.character(sample_number)) %>%
      mutate(sample_name=gsub("RESUB-", "", sample_name))) %>%
  mutate(sample_number=gsub(".G|.S", "", sample_number))%>% 
  group_by(sample_number) %>% fill(c(tank, condition), .direction = "down") %>% #fill in missing tank and treatment info 
  mutate(tissue_type=case_when(
    grepl(".G", sample_name) ~ "Gill",
    grepl(".S", sample_name) ~ "Spleen",
    TRUE ~ "Liver") %>% as.factor()) %>%
  column_to_rownames("sample_name")

samp.0.9 <- samp %>% filter(temp_treatment %in% c(0,9)) %>% droplevels()
samp.5.9 <- samp %>% filter(temp_treatment %in% c(5,9)) %>% droplevels()
samp.16.9 <- samp %>% filter(temp_treatment %in% c(16,9)) %>% droplevels()
```

Write out more featurecounts-style exon count matrices with subsets of samples for separate pairwise contrasts: 

```{r}
write_delim(exon.counts.fc.s[,c("Geneid", "Chr", "Start", "End", "Strand", "Length", rownames(samp.0.9))], 
            file = "../aligned/featurecounts_exon_dexseq_filtered_0.9", delim = "\t", col_names = T)
write_delim(exon.counts.fc.s[,c("Geneid", "Chr", "Start", "End", "Strand", "Length", rownames(samp.5.9))], 
            file = "../aligned/featurecounts_exon_dexseq_filtered_5.9", delim = "\t", col_names = T)
write_delim(exon.counts.fc.s[,c("Geneid", "Chr", "Start", "End", "Strand", "Length", rownames(samp.16.9))], 
            file = "../aligned/featurecounts_exon_dexseq_filtered_16.9", delim = "\t", col_names = T)
```

```{r}
# All samples/treatments, to identify DUEs in any non-optimal treatment 
dxd.fc <- DEXSeqDataSetFromFeatureCounts(countfile = "../aligned/featurecounts_exon_dexseq_filtered",
                                         flattenedfile = "../../references/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf", 
                                         sampleData = samp %>% dplyr::select(condition))
dxd.fc.0vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "../aligned/featurecounts_exon_dexseq_filtered_0.9",
                                         flattenedfile = "../../references/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf", 
                                         sampleData = samp.0.9 %>% dplyr::select(condition))

dxd.fc.5vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "../aligned/featurecounts_exon_dexseq_filtered_5.9",
                                         flattenedfile = "../../references/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf", 
                                         sampleData = samp.5.9 %>% dplyr::select(condition))

dxd.fc.16vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "../aligned/featurecounts_exon_dexseq_filtered_16.9",
                                         flattenedfile = "../../references/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf", 
                                         sampleData = samp.16.9 %>% dplyr::select(condition))

```
## DEXSeq Analysis

### Explore data 

```{r}
colData(dxd.fc)
```
```{r}
head(counts(dxd.fc), 2)
```
```{r}
head( featureCounts(dxd.fc), 3 )
```

```{r}
head(rowRanges(dxd.fc), 5)
```
```{r}
sampleAnnotation(dxd.fc)
```
### Normalisation 

```{r}
dxd.fc = estimateSizeFactors( dxd.fc )
```

### Dispersion estimation 
```{r}
# BPPARAM = BiocParallel::SnowParam(6) #use "SnowParam" on Windows, but "MulticoreParam" on Mac
# dxd.fc = DEXSeq::estimateDispersions( dxd.fc, BPPARAM=BPPARAM )
# save(dxd.fc)
# plotDispEsts(dxd.fc)
```

### Test for differential exon usage 

```{r}
# dxd.fc = testForDEU( dxd.fc )
# dxd.fc = estimateExonFoldChanges( dxd.fc, fitExpToVar="condition")
```

### Access results from DEXSeq analysis
```{r}
dxr1 = DEXSeqResults( dxd.fc )
dxr1
load("../../rnaseq/dexseq/dxr1")
```

```{r}
mcols(dxr1)$description
dxr1
```
How many exonic regions are significant at FDR 10%? FDR 5%?

```{r}
table ( dxr1$padj < 0.05 )
```
How many genes are affected?
```{r}
table ( tapply( dxr1$padj < 0.05, dxr1$groupID, any ) )
```
```{r}
plotMA( dxr1, cex=0.6 )
```
```{r}
#table ( tapply( dxr1$padj < 0.05, dxr1$groupID, any ) )
mcols(dxr1)$description
```

```{r}
load(file = "../deseq2/diffex.0.filt")
load(file = "../deseq2/diffex.5.filt")
load(file = "../deseq2/diffex.16.filt")

diffex.0.names <- (diffex.0.filt %>% as.data.frame() %>% rownames_to_column("ncbi_id") %>% mutate(ncbi_id=gsub("GeneID:", "", ncbi_id)) %>%
left_join(pcod.gff %>% filter(type=="gene")))$name

diffex.5.names <- (diffex.5.filt %>% as.data.frame() %>% rownames_to_column("ncbi_id") %>% mutate(ncbi_id=gsub("GeneID:", "", ncbi_id)) %>%
left_join(pcod.gff %>% filter(type=="gene")))$name

diffex.16.names <- (diffex.16.filt %>% as.data.frame() %>% rownames_to_column("ncbi_id") %>% mutate(ncbi_id=gsub("GeneID:", "", ncbi_id)) %>%
left_join(pcod.gff %>% filter(type=="gene")))$name
    
dxr1.details <- dxr1 %>% as.data.frame() %>% filter(padj<0.05) %>% filter(groupID %!in% c(diffex.0.names, diffex.5.names, diffex.16.names)) %>%
#  mutate(largest_fc=pmax(abs(log2fold_b_0_a_9), abs(log2fold_c_5_a_9), abs(log2fold_d_16_a_9)))
  pivot_longer(cols = c(log2fold_b_0_a_9, log2fold_c_5_a_9, log2fold_d_16_a_9), names_to = "contrast", values_to = "LFC") %>% 
  mutate(exposure=case_when(
    contrast=="log2fold_b_0_a_9"~"Cold",
    contrast=="log2fold_c_5_a_9"~"Cool",    
    contrast=="log2fold_d_16_a_9"~"Warm")) %>%
  dplyr::select(groupID, featureID, exonBaseMean, dispersion, stat, pvalue, padj, contrast, exposure, LFC, everything()) %>%
    group_by(groupID, featureID) %>% summarise_each(funs(.[which.max(abs(LFC))]))
```
```{r}
dxr1

warm.top.exon.genes <- (dxr1.details %>% filter(exposure=="Warm") %>% arrange(padj) %>% head(n=10))$groupID
cool.top.exon.genes <- (dxr1.details %>% filter(exposure=="Cool") %>% arrange(padj) %>% head(n=10))$groupID
cold.top.exon.genes <- (dxr1.details %>% filter(exposure=="Cold") %>% arrange(padj) %>% head(n=10))$groupID

for (i in 1:10) {
  print(plotDEXSeq( dxr1, cold.top.exon.genes[i], expression=TRUE, displayTranscripts = F, norCounts=FALSE, legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 ))
}

```


```{r}
dxr1 %>% as.data.frame() %>%
  filter(groupID == "LOC132464016")

plotDEXSeq( dxr1, "LOC132464016", expression=TRUE, displayTranscripts = TRUE, norCounts=FALSE, legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 )
plotDEXSeq( dxr1, "LOC132464016", expression=FALSE, displayTranscripts = TRUE, norCounts=TRUE, legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2 )
```
### Rerun DEXSeq on subsetted objects, one for each treatment (0, 5, and 16, all vs. 9)

The following code was run on Sedna using the scripts [DEXSeq.R]() (executed using the slurm script [dexseq.sh]())

```{r}
# Rerun DEXSeq on subsetted objects, one for each treatment (0, 5, and 16, all vs. 9)

dxd.fc <- DEXSeqDataSetFromFeatureCounts(countfile = "featurecounts_exon_dexseq_filtered",
                                         flattenedfile = "/home/lspencer/references/pcod-ncbi/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf",
                                         sampleData = samp %>% dplyr::select(condition))


dxd.fc.0vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "featurecounts_exon_dexseq_filtered_0.9",
                                         flattenedfile = "/home/lspencer/references/pcod-ncbi/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf",
                                         sampleData = samp.0.9 %>% dplyr::select(condition))

dxd.fc.5vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "featurecounts_exon_dexseq_filtered_5.9",
                                         flattenedfile = "/home/lspencer/references/pcod-ncbi/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf",
                                         sampleData = samp.5.9 %>% dplyr::select(condition))

dxd.fc.16vs9 <- DEXSeqDataSetFromFeatureCounts(countfile = "featurecounts_exon_dexseq_filtered_16.9",
                                         flattenedfile = "/home/lspencer/references/pcod-ncbi/GCF_031168955.1_ASM3116895v1_genomic_flat.gtf",
                                         sampleData = samp.16.9 %>% dplyr::select(condition))

dxd.fc.treats <- vector(mode = "list", length = 3)
names(dxd.fc.treats) <-c ("dxd.fc.0vs9", "dxd.fc.5vs9", "dxd.fc.16vs9")
dxd.fc.treats[1] <- dxd.fc.0vs9
dxd.fc.treats[2] <- dxd.fc.5vs9
dxd.fc.treats[3] <- dxd.fc.16vs9
dxr.treats <- vector(mode = "list", length = 3)
names(dxr.treats) <- c("dxr.0vs9", "dxr.5vs9", "dxr.16vs9")

for (i in 2:3){

 # Normalization
 dxd.fc.treats[[i]] = estimateSizeFactors( dxd.fc.treats[[i]] )

 # Dispersion estimation
 ## This is very memory intensive. Here I use multiple cores.
 BPPARAM = BiocParallel::MulticoreParam(24)
 dxd.fc.treats[[i]] = DEXSeq::estimateDispersions( dxd.fc.treats[[i]], BPPARAM=BPPARAM )

 pdf(file = paste("disp-estimates_", gsub("dxd.fc.", "", names(dxd.fc.treats[i])),".pdf", sep=""), width = 12, height = 9);
 plotDispEsts( dxd.fc.treats[[i]] )
 dev.off()

 dxd.fc.treats[[i]] = testForDEU( dxd.fc.treats[[i]], BPPARAM=BPPARAM )

 # Estimate fold changes
 dxd.fc.treats[[i]] = estimateExonFoldChanges( dxd.fc.treats[[i]], fitExpToVar="condition", BPPARAM=BPPARAM )

 # Extract results object
 dxr.treats[[i]] = DEXSeqResults( dxd.fc.treats[[i]] )

 pdf(file = paste("volcano-plot_", gsub("dxd.fc.", "", names(dxd.fc.treats[i])),".pdf", sep=""), width = 12, height = 9);
 plotMA(  dxr.treats[[i]], cex=0.8 )
 dev.off()

 save(dxd.fc.treats, file="dxd.fc.treats")
 save(dxr.treats, file="dxr.treats")
 }

# Save again, just in case!
 save(dxd.fc.treats, file="dxd.fc.treats")
 save(dxr.treats, file="dxr.treats")
```

Load in results object, which is a list that contains differential exon usage results for each of the three treatments: 

```{r}
load(file = "../dexseq/dxr.treats") #DEU results list, containing results for each 9vs. contrast 

for (i in 1:length(dxr.treats)){
  print(names(dxr.treats[i]))
  print(paste("Number of differentially used exons: ", dxr.treats[[i]] %>% as.data.frame() %>% filter(padj<0.05) %>% nrow(), sep=""))
  print(paste("Number of genes affected: ", dxr.treats[[i]] %>% as.data.frame() %>% filter(padj<0.05) %>% 
                dplyr::select(groupID) %>% distinct() %>% nrow(), sep=""))
}

numbOfGenes = sum( perGeneQValue(dxr.treats$dxr.16vs9) < 0.05)

```
### Let's explore results from the 9C vs. 16C contrast. 

```{r}
plotMA(dxr.treats$dxr.16vs9, cex=0.5, alpha=0.05, ylim=c(-5,5), xlim=(c(1, 1e6)),
       main="Differenitally used exons, 16C vs. control (9C)")
```
Let's look at top 20 genes, i.e. 20 genes containin exons with the lowest padj values (or highest |log2FC|, whichever you choose) 

```{r}
dxr.16.top10genes <- (dxr.treats$dxr.16vs9 %>% as.data.frame() %>% filter(padj<0.05 & abs(log2fold_d_16_a_9)>1) %>% 
                        arrange(padj) %>%  #sort by padj
                        #arrange(desc(abs(log2fold_d_16_a_9))) %>% # sort by abs(l2fc) 
                        head(n=20))$groupID

for (i in 1:length(dxr.16.top10genes)) {
  plotDEXSeq(dxr.treats$dxr.16vs9, dxr.16.top10genes[i], displayTranscripts=TRUE, 
             splicing=TRUE, expression=FALSE, #here I opt to remove differences in overall expression and highlight differences in exon usage/expression
             legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2, color = c("green4", "firebrick2"))
}
```
```{r}
#DEXSeqHTML(dxr.treats$dxr.16vs9, FDR=0.05, color=c("green4", "firebrick2"))
```

DAVID enrichment analysis

```{r}
deu.16.annot <- dxr.treats$dxr.16vs9 %>% as.data.frame() %>% filter(padj<0.05 & abs(log2fold_d_16_a_9)>.5) %>% 
  left_join(pcod.blast.GO, by=c("groupID"="gene_pcod")) %>%
  arrange(padj)

require(clipr)

(deu.16.annot %>%
  remove_rownames() %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
  filter(!is.na(spid)))$spid %>%
  write_clip()

# All genes
(dxr.treats$dxr.16vs9 %>% as.data.frame() %>% 
  left_join(pcod.blast.GO, by=c("groupID"="gene_pcod")) %>%
    remove_rownames() %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
    filter(!is.na(spid)))$spid %>%
  write_clip()

david.deu.16 <- data.frame(read_delim(file="../dexseq/DAVID_DEU_16C.txt", delim = "\t")) %>% 
  clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

save(david.deu.16, file = "../dexseq/david.deu.16")
require(ggpubr)

david.deu.16 %>%
  #filter(count>=3) %>% 
  filter(p_value<0.01) %>%
  #filter(fdr<0.1) %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.deu.16 %>% 
              dplyr::select(category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%

  ungroup() %>% arrange(p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  add_column("stress"="Warm") %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")
```
```{r, warning=F, message=F}
require(wordcloud)
# Generate wordcloud from GO terms 

# # All GO terms associated with all DEU genes
# deu.16.annot %>% 
#   dplyr::select(groupID, featureID, padj,log2fold_d_16_a_9, ncbi_id, gene_uni, spid, evalue, protein_names, gene_ontology_biological_process) %>%
#   select(gene_ontology_biological_process) %>% 
#   #select(gene_ontology_cellular_component) %>%
#   #select(protein_names) %>% 
#   na.omit() %>% unlist() %>% as.vector() %>% 

# Enriched GO terms  
  david.deu.16 %>%
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
  #filter(fdr<0.1) %>%
  filter(category %in% c("GOTERM_BP_DIRECT", "UP_KW_BIOLOGICAL_PROCESS")) %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
    select(process) %>% 
  na.omit() %>% unlist() %>% as.vector() %>% 

  
#   gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
#   gsub("interleukin-..|interleukin-.|interleukin.", "interleukin", ., perl=TRUE) %>%  #gsub("process|regulation|negative|positive|response|pathway|production", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
 #gsub("process|regulation|positive|negative|response|pathway|cell|protein", "", ., perl=TRUE) %>% #get rid of generic words that are in everything
  wordcloud::wordcloud(min.freq = 1, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))


```

Re-run DAVID without genes that were identified as differentially expressed

```{r}

(deu.16.annot %>%
    filter(groupID %!in% rownames(diffex.16.filt)) %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
  filter(!is.na(spid)))$spid %>%
  write_clip()

# All genes
(dxr.treats$dxr.16vs9 %>% as.data.frame() %>% 
  left_join(pcod.blast.GO, by=c("groupID"="gene_pcod")) %>%
    remove_rownames() %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
    filter(!is.na(spid)))$spid %>%
  write_clip()

david.deu.noDEGs.16 <- data.frame(read_delim(file="../dexseq/DAVID_DEU-noDEGs_16C.txt", delim = "\t")) %>% 
  clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

david.deu.noDEGs.16 %>%
  #filter(count>=3) %>% 
  filter(p_value<0.01) %>%
  #filter(fdr<0.1) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.deu.noDEGs.16 %>% 
              dplyr::select(category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%

  ungroup() %>% arrange(p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  add_column("stress"="Warm") %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")
```
```{r}
# Enriched GO terms  
  david.deu.noDEGs.16 %>%
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
  #filter(fdr<0.1) %>%
  filter(category %in% c("GOTERM_BP_DIRECT", "UP_KW_BIOLOGICAL_PROCESS")) %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
    select(process) %>% 
  na.omit() %>% unlist() %>% as.vector() %>% 

  
#   gsub("\\[[^\\]]*\\]", "", ., perl=TRUE) %>%
#   gsub("interleukin-..|interleukin-.|interleukin.", "interleukin", ., perl=TRUE) %>%  #gsub("process|regulation|negative|positive|response|pathway|production", "", ., perl=TRUE) %>% #get rid of word "process" since it's in everything
 #gsub("process|regulation|positive|negative|response|pathway|cell|protein", "", ., perl=TRUE) %>% #get rid of generic words that are in everything
  wordcloud::wordcloud(min.freq = 1, random.order = F, random.color = F, colors = brewer.pal(2, "Dark2"))
```

```{r}
deu.16.annot %>% dplyr::select(!contains("countData")) %>% 
  dplyr::select(groupID, padj, log2fold_d_16_a_9, evalue, protein_names ) %>% 
  dplyr::select(groupID, protein_names) %>% distinct()
```

Venn diagram of genes containing differentially expressed exons or transcripts from my analysis, Sam's, and Steven's 

```{r, warning=F, message=F}
DET.steven <- read_delim("../DETlist_steven.tab", delim = "\t", skip = 1, 
                         col_names=c("transcript_id", colnames(read_delim("../DETlist_steven.tab")))) %>%
  as.data.frame() %>% mutate(transcript_id=gsub("rna-", "", transcript_id)) %>%
  left_join(pcod.gtf.transcript2gene %>% dplyr::select(transcript_id, gene_id, ncbi_id)) %>% 
  mutate(gene_id=case_when(is.na(gene_id) ~ transcript_id, TRUE ~ gene_id)) %>% 
  mutate(gene_id=gsub("gene-|id-", "", gene_id))
  
DEG.steven <- read_delim("../DEGlist_steven.tab", delim = "\t", skip = 1, 
                         col_names=c("gene_id", colnames(read_delim("../DEGlist_steven.tab")))) %>%
  as.data.frame() %>% mutate(gene_id=gsub(".*\\|","",gene_id))

DET.sam <- read_delim("../DET_temp_pvalOnly_filtered_stat_results_p-0.05.csv") %>% as.data.frame()
DEG.sam <- read_delim("../DEG_temp_pvalOnly_filtered_stat_results_p-0.05.csv") %>% as.data.frame() %>%
              mutate(geneNames=gsub("gene-", "", id))
DEU.laura <- deu.16.annot
DEG.laura <- diffex.16.filt %>% as.data.frame() %>% rownames_to_column("gene_id")

DEG.kathleen <- read_delim("../Gmac_DEGs_sig_L.9.16_norm.tab")[-1] %>%
  left_join(pcod.blast.GO, by=c("gene"="gene_pcod"))
```

```{r}
`DEGs (Laura)` <- DEG.laura$gene_id %>% unique()
`DEGs (Sam)`  <- DEG.sam$geneNames %>% unique()
`DEGs (Steven)` <- DEG.steven$gene_id %>% unique()
`DEGs (Kathleen)` <- DEG.kathleen$gene %>% unique()

# All DEU/Ts
`DEUs (Laura)` <- DEU.laura$groupID %>% unique()
`DETs (Sam)`  <- DET.sam$geneNames %>% unique()
`DETs (Steven)` <- DET.steven$gene_id %>% unique()

# DEU/Ts that do not include DEGs
`DEUs (Laura)` <- (DEU.laura %>% filter(groupID %!in% DEG.laura$gene_id))$groupID %>% unique()
`DETs (Sam)`  <- (DET.sam %>% filter(geneNames %!in% DEG.sam$geneNames))$geneNames %>% unique()
`DETs (Steven)` <- (DET.steven %>% filter(gene_id %!in% DEG.steven$gene_id))$gene_id %>% unique()

venn.alt.splice <- data.frame(dat = 
             unique(c(`DEUs (Laura)`,
                      `DETs (Steven)`,
                      `DETs (Sam)`))) %>%
  mutate(`DEUs (Laura)` = dat %in% `DEUs (Laura)`,
         `DETs (Steven)` = dat %in% `DETs (Steven)`,
         `DETs (Sam)` = dat %in% `DETs (Sam)`) %>%
  dplyr::select(`DEUs (Laura)`,
                `DETs (Steven)`,
                `DETs (Sam)`) %>% euler()

venn.degs <- data.frame(dat = 
             unique(c(`DEGs (Laura)`,
                      `DEGs (Steven)`,
#                      `DEGs (Sam)`,
                      `DEGs (Kathleen)`))) %>%
  mutate(`DEGs (Laura)` = dat %in% `DEGs (Laura)`,
         `DEGs (Steven)` = dat %in% `DEGs (Steven)`,
#         `DEGs (Sam)` = dat %in% `DEGs (Sam)`,
         `DEGs (Kathleen)` = dat %in% `DEGs (Kathleen)`) %>%
  dplyr::select(`DEGs (Laura)`,
                `DEGs (Steven)`,
#                `DEGs (Sam)`,
                `DEGs (Kathleen)`) %>% euler()


venn.laura <- data.frame(dat = 
             unique(c(`DEUs (Laura)`,
                      `DEGs (Laura)`))) %>%
  mutate(`DEUs (Laura)` = dat %in% `DEUs (Laura)`,
         `DEGs (Laura)` = dat %in% `DEGs (Laura)`) %>%
  dplyr::select(`DEUs (Laura)`,
                `DEGs (Laura)`) %>% euler()
```

Venn diagrams to look at overlap among approaches

Laura = STAR --> featureCounts --> DEXSeq
Sam= HiSAT --> Stringtie --> Ballgown
Steven= HiSAT --> Stringtie --> DESeq 

```{r}
venns <-   list(venn.alt.splice, venn.degs) 
names(venns) <- c("Alternative Splicing", "Differential Gene Expression")
for (i in 1:length(venns)) {
  print(eulerr:::plot.euler(venns[[i]],
  counts = T, quantities = T, 
                      legend = T, edges=F, fontzie=8,
                      font=1, cex=1, alpha=0.5, 
                      brewer.pal(10, "Dark2"), main=names(venns[i])))
}

# Overlapping genes containing DEUs (laura) and DETs (Sam)
Reduce(intersect, list(`DEUs (Laura)`, `DETs (Steven)`, `DETs (Sam)`)) %>%
  as.data.frame() %>% `colnames<-`("gene_pcod") %>%
  left_join(pcod.blast.GO) %>% 
  dplyr::select(gene_pcod, spid, description ) %>% write_clip()
#  filter(!is.na(spid)) %>% dplyr::select(spid) %>% unlist() %>% write_clip()
  
alt.splice.overlap <- Reduce(intersect, list(`DEUs (Laura)`, `DETs (Steven)`, `DETs (Sam)`)) %>% unique()
for (i in 1:length(alt.splice.overlap)) {
  plotDEXSeq(dxr.treats$dxr.16vs9, alt.splice.overlap[i], displayTranscripts=FALSE, 
             splicing=TRUE, expression=FALSE, #here I opt to remove differences in overall expression and highlight differences in exon usage/expression
             legend=TRUE, cex.axis=1.2, cex=1.3, lwd=2, color = c("green4", "firebrick2"))
}
```

GO enrichment of genes that contain DETs

```{r}
# DAVID with Sam's genes containing DETs 
DET.sam %>%
  left_join(pcod.blast.GO, by=c("geneNames"="gene_pcod")) %>%
  filter(!is.na(spid)) %>% dplyr::select(spid) %>% unlist() %>% write_clip()

DET.steven %>% left_join(pcod.blast.GO %>% dplyr::select(-ncbi_id), by=c("gene_id"="gene_pcod")) %>%
  filter(!is.na(spid)) %>% dplyr::select(spid) %>% unlist() %>% write_clip()

# All genes
(dxr.treats$dxr.16vs9 %>% as.data.frame() %>% 
  left_join(pcod.blast.GO, by=c("groupID"="gene_pcod")) %>%
    remove_rownames() %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
    filter(!is.na(spid)))$spid %>%
  write_clip()

```


```{r}
#david.DETs.sam <- data.frame(read_delim(file="../david_sam-DEUs.txt", delim = "\t")) %>% 
david.DETs.steven <- data.frame(read_delim(file="../david_steven-DETs.txt", delim = "\t")) %>% 
  clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

#david.DETs.sam %>%
david.DETs.steven %>%
  #filter(count>=3) %>% 
  filter(p_value<0.01) %>%
  #filter(fdr<0.1) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.DETs.steven %>% 
              dplyr::select(category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%

  ungroup() %>% arrange(p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  add_column("stress"="Warm") %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")

save(david.DETs.sam, file = "../david.DETs.sam")
save(david.DETs.steven, file = "../david.DETs.steven")
getwd()
```

```{r}
# Overlapping DEGs (laura) and DETs (Sam)
Reduce(intersect, list(rownames(diffex.16.filt), DEG.sam)) %>%
  as.data.frame() %>% `colnames<-`("gene_pcod") %>%
  left_join(pcod.blast.GO) %>%
  filter(!is.na(spid)) %>% dplyr::select(spid) %>% unlist() %>% write_clip()

length(DEUs.laura)

Reduce(intersect, list(rownames(diffex.16.filt), DEUs.laura)) %>% length() 
```

GO enrichment of genes that contain both DEUs (Laura) and DETs (Sam)

```{r}
david.alt.splice <- data.frame(read_delim(file="../david_DEUs-DETs.txt", delim = "\t")) %>% 
  clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

david.alt.splice %>%
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
  #filter(fdr<0.1) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.alt.splice %>% 
              dplyr::select(category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%

  ungroup() %>% arrange(p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  add_column("stress"="Warm") %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")
```
GO enrichment of overlapping DEGs (Laura,Sam)

```{r}
david.overlap.degs <- data.frame(read_delim(file="../david_overtlapping-degs.txt", delim = "\t")) %>% 
  clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~")

david.overlap.degs %>%
  #filter(count>=3) %>% 
  filter(p_value<0.05) %>%
  #filter(fdr<0.1) %>%
  filter(category=="GOTERM_BP_DIRECT") %>% 
#  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%

  group_by(genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.overlap.degs %>% 
              dplyr::select(category, term, process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(stress, response, p_value) %>% group_by(stress, response) %>% dplyr::slice(1:15) %>%

  ungroup() %>% arrange(p_value) %>%
  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
  add_column("stress"="Warm") %>%

ggplot(aes(y = process, x=stress, col=stress)) + 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
 scale_color_manual(name="Temperature", values=c("Cold"="royalblue1","Cool"="darkgreen", "Warm"="firebrick4"),
                                guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle("Enriched Biological Processes")
```

Conduct functional enrichment for each contrast using goseq (Test out goseq using Ariana's code)


```{r}
BiocManager::install("goseq", version = "3.18")

(deu.16.annot %>%
    filter(groupID %!in% rownames(diffex.16.filt)) %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
  filter(!is.na(spid)))$spid %>%
  write_clip()

# All genes
(dxr.treats$dxr.16vs9 %>% as.data.frame() %>% 
  left_join(pcod.blast.GO, by=c("groupID"="gene_pcod")) %>%
    remove_rownames() %>%
  dplyr::select(groupID, spid) %>% distinct() %>%
    filter(!is.na(spid)))$spid %>%
  write_clip()


### Generate vector with names of all genes detected in our dataset 
ALL.vector <- c(filtered_Mcap.annot$gene)

### Generate length vector for all genes 
LENGTH.vector <- as.integer(filtered_Mcap.annot$length)

### Generate vector with names in just the contrast we are analyzing
ID.vector <- temperature_up_list%>%
  filter(contrast==temp)%>%
  pull(gene)

##Get a list of GO Terms for all genes detected in our dataset 
GO.terms <- filtered_Mcap.annot%>%
  dplyr::select(gene, GOs)

##Format to have one goterm per row with gene ID repeated
split <- strsplit(as.character(GO.terms$GOs), ";") 
split2 <- data.frame(v1 = rep.int(GO.terms$gene, sapply(split, length)), v2 = unlist(split)) 
colnames(split2) <- c("gene", "GOs")
GO.terms<-split2

##Construct list of genes with 1 for genes in contrast and 0 for genes not in the contrast
gene.vector=as.integer(ALL.vector %in% ID.vector) 
names(gene.vector)<-ALL.vector#set names

#weight gene vector by bias for length of gene 
pwf<-nullp(gene.vector, ID.vector, bias.data=LENGTH.vector) 

#run goseq using Wallenius method for all categories of GO terms 
GO.wall<-goseq(pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:BP"), method="Wallenius", use_genes_without_cat=TRUE)

GO <- GO.wall[order(GO.wall$over_represented_pvalue),]
colnames(GO)[1] <- "GOterm"

#adjust p-values 
GO$bh_adjust <-  p.adjust(GO$over_represented_pvalue, method="BH") #add adjusted p-values

#Filtering for p < 0.01
GO <- GO %>%
    dplyr::filter(bh_adjust<0.05) %>%
    dplyr::arrange(., ontology, bh_adjust)

#Write file of results 
write_csv(GO, file = paste0("output/rna_seq/functional_enrichment/goseq_temperature_up_", temp, ".csv"))

```

```

