---
title: "04-WGCNA"
author: "Laura H Spencer"
date: "4/29/2024"
output: html_document
---

```{r, message=FALSE, warning=FALSE, results=FALSE}
# Add all required libraries that are installed with install.packages() here
list.of.packages <- c("tidyverse", "RColorBrewer", "ggpmisc", "ggpubr", "ape", "pheatmap", "plotly", "clipr")
# Add all libraries that are installed using BiocManager here
bioconductor.packages <- c("WGCNA", "DESeq2")

# Install BiocManager if needed
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

# Get names of all required packages that aren't installed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[, "Package"])]
new.bioc.packages <- bioconductor.packages[!(bioconductor.packages %in% installed.packages()[, "Package"])]
# Install all new packages
if(length(new.packages)) install.packages(new.packages)
if(length(new.bioc.packages)) BiocManager::install(new.bioc.packages)

# Load all required libraries
all.packages <- c(list.of.packages, bioconductor.packages)
lapply(all.packages, FUN = function(X) {
  do.call("require", list(X))
})
```

Format of expression data should be: 
- rownames = sample 
- columns = genes 
- normalized 

# Load in filtered datasets also used for DESeq2 

```{r}
# load data
load(file = "../deseq2/counts.trans") 
load(file = "../deseq2/vsd.temperature")

# Read in sample metadata
load(file="../sample.info.rna")
```

```{r}
# Load annotation objects 
load(file = "../../references/Pcod.blast")
load(file = "../../references/Pcod.blast.GO")
load(file = "../aligned/counts.annot.pcod")

# load in DEG lists 
load(file = "../deseq2/diffex.0.filt")
load(file = "../deseq2/diffex.5.filt")
load(file = "../deseq2/diffex.16.filt")

# Generate list of genes that are DEGs 
degs <- unique(c(rownames(diffex.0.filt),rownames(diffex.5.filt), rownames(diffex.16.filt)))

# Load in DESeq2 results lists to pull background list of genes 
load(file = "../deseq2/res.0") #res.temperature.AM
load(file = "../deseq2/res.5") #res.temperature.AL
load(file = "../deseq2/res.16") #res.temperature.ML
```

```{r}
#allowWGCNAThreads()
enableWGCNAThreads()
```

```{r}
# Use WGCNA's `goodSamplesGenes` to remove any genes with very low variability or missing data  
datExpr0 <- counts.trans[,-1:-2]
#datExpr0 <- counts.trans[,3:1000] #for testing purposes use subset of data

gsg = goodSamplesGenes(datExpr0, verbose = 3);
sum(gsg$goodGenes==FALSE)  #were any removed?  Nope. 
ncol(gsg$allOK)

if (!gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0) 
     printFlush(paste("Removing genes:", paste(names(datExpr0)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0) 
     printFlush(paste("Removing samples:", paste(rownames(datExpr0)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr0 = datExpr0[gsg$goodSamples, gsg$goodGenes]
}
```

```{r}
sampleTree = hclust(dist(datExpr0), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
# The user should change the dimensions if the window is too large or too small.
sizeGrWindow(12,9)

#pdf(file = "/home/lspencer/2022-redking-OA/wgcna/sampleClustering.pdf", width = 12, height = 9);
#pdf(file = "C:/Users/laura.spencer/Desktop/sampleClustering.pdf", width = 12, height = 9);
par(cex = 0.6);
par(mar = c(0,4,2,0))
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5, 
     cex.axis = 1.5, cex.main = 2)
# Plot a line to show the cut
abline(h = 128, col = "red");
#dev.off()

# Determine cluster under the line
clust = cutreeStatic(sampleTree, cutHeight = 128, minSize = 10)
table(clust)
# clust 1 contains the samples we want to keep.
keepSamples = (clust==1)
#keepSamples = (clust==0)
datExpr = datExpr0[keepSamples, ]
nGenes = ncol(datExpr)
nSamples = nrow(datExpr)

# Here I removed samples 48 & 98; both of those had low RNA levels and DV200<70 during QC 
```

```{r}
# traitData = read.csv("wgcna-testing/ClinicalTraits.csv");
# dim(traitData)
# names(traitData)

# remove columns that hold information we do not need.
(allTraits = counts.trans[, 1:2] %>% rownames_to_column("Sample") %>%
   mutate(temp_num=as.numeric(as.character(temperature))))

dim(allTraits)
names(allTraits)

# # Form a data frame analogous to expression data that will hold the traits.
# 
Samples = rownames(datExpr);
traitRows = match(Samples, allTraits$Sample);
datTraits = allTraits[traitRows, c(3, 4)];
rownames(datTraits) = allTraits[traitRows, 1];

# 
collectGarbage();
```

```{r}
# Re-cluster samples
sampleTree2 = hclust(dist(datExpr), method = "average")
# Convert traits to a color representation: white means low, red means high, grey means missing entry
traitColors = numbers2colors(sapply(datTraits,as.numeric), signed = F, colors = colorRampPalette(brewer.pal(8, "Set2"))(14));

# Plot the sample dendrogram and the colors underneath.
#pdf(file = "C:/Users/laura.spencer/Desktop/sampleTree.pdf", width = 12, height = 9);
plotDendroAndColors(sampleTree2, traitColors,
                    groupLabels = names(datTraits), 
                    main = "Sample dendrogram with treatment and tank ids")
#dev.off()
```

```{r}
options(stringsAsFactors = FALSE);
```

"Constructing a weighted gene network entails the choice of the soft thresholding power Î² to which co-expression
similarity is raised to calculate adjacency. The authors of [1] have proposed to choose the soft thresholding power based on the criterion of approximate scale-free topology." 

I will choose the power 9, which is the lowest power for which the scale-free topology fit
index reaches 0.90.

```{r}
# Choose a set of soft-thresholding powers
powers = c(c(1:25))
# Call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5, networkType = "signed")
# Plot the results:
sizeGrWindow(9, 5)
#pdf(file = "C:/Users/laura.spencer/Desktop/soft-thresholding power.pdf", width = 12, height = 9);
par(mfrow = c(1,2));
cex1 = 0.9;
# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col="red")

# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
#dev.off()
```

I now calculate the adjacencies, using the soft thresholding power 9:

```{r}
softPower = 15;
adjacency = adjacency(datExpr, power = softPower, type="signed", corFnc="bicor");

# for testing purposes, subset expression dataframe 
#adjacency = adjacency(datExpr[,1:2000], power = softPower, type="signed", corFnc="bicor");
```

To minimize effects of noise and spurious associations, we transform the adjacency into Topological Overlap Matrix, and calculate the corresponding dissimilarity:

```{r}
TOM = TOMsimilarity(adjacency); #this takes a very long time
dissTOM = 1-TOM

save(dissTOM, file = "../wgcna/dissTom")
#load(file = "../wgcna/dissTom") #load if needed
```

We now use hierarchical clustering to produce a hierarchical clustering tree (dendrogram) of genes. Note that we use the function hclust that provides a much faster hierarchical clustering routine than the standard hclust function.

```{r}
# Call the hierarchical clustering function
geneTree = hclust(as.dist(dissTOM), method = "average");
save(file = "../wgcna/geneTree")
# Plot the resulting clustering tree (dendrogram)

#load(file = "../wgcna/geneTree")

sizeGrWindow(12,9)
#pdf(file = "C:/Users/laura.spencer/Desktop/gene-clustering.pdf", width = 12, height = 9);
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
     labels = FALSE, hang = 0.04);
#dev.off()
```

```{r}
# We like large modules, so we set the minimum module size relatively high:
minModuleSize = 30;

#Module identification using dynamic tree cut:
dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM,
                deepSplit = 2, pamRespectsDendro = FALSE,
                minClusterSize = minModuleSize);

# dynamicMods.50 = cutreeDynamic(dendro = geneTree, distM = dissTOM,
#                 deepSplit = 2, pamRespectsDendro = FALSE,
#                 minClusterSize = 50);
# dynamicMods.75 = cutreeDynamic(dendro = geneTree, distM = dissTOM,
#                 deepSplit = 2, pamRespectsDendro = FALSE,
#                 minClusterSize = 75);

#load(file = "../wgcna/dynamicMods")
#load(file = "../wgcna/dynamicMods.50")
#load(file = "../wgcna/dynamicMods.75")

table(dynamicMods)
#table(dynamicMods.50)
#table(dynamicMods.75)  # final analysis uses minimum module size 75 (so many genes!)
```

```{r}
# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
#pdf(file = "C:/Users/laura.spencer/Desktop/gene-clustering-modules.pdf", width = 12, height = 9);
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene clustering on TOM-based dissimilarity\nDendrogram and module colors")
#dev.off()
```

The Dynamic Tree Cut may identify modules whose expression profiles are very similar. It may be prudent to merge
such modules since their genes are highly co-expressed. To quantify co-expression similarity of entire modules, we
calculate their eigengenes and cluster them on their correlation.

We choose a height cut of 0.25, corresponding to correlation of 0.75, to merge. 


```{r}
# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");

# Plot the result
sizeGrWindow(7, 6)
#pdf(file = "C:/Users/laura.spencer/Desktop/module-tree.pdf", width = 12, height = 9);
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")
MEDissThres = 0.25
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
#dev.off()

# Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;
```


```{r}
#sizeGrWindow(12, 9)
#pdf(file = "C:/Users/laura.spencer/Desktop/gene-clustering-modules-merged.pdf", width = 12, height = 9);
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                    c("Dynamic Tree Cut", "Merged dynamic"),
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05,
                    main = "Gene clustering on TOM-based dissimilarity\nDendrogram and module colors")
#dev.off()

# sizeGrWindow(12, 9)
# plotDendroAndColors(geneTree, mergedColors,
#                     c("Dynamic Tree Cut", "Merged dynamic"),
#                     dendroLabels = FALSE, hang = 0.03,
#                     addGuide = TRUE, guideHang = 0.05,
#                     main = "Gene clustering on TOM-based dissimilarity\nFinalized modules")
```



```{r}
# Rename to moduleColors
moduleColors = mergedColors
# Construct numerical labels corresponding to the colors
colorOrder = c("grey", standardColors(50));
moduleLabels = match(moduleColors, colorOrder)-1;
MEs = mergedMEs;
# Save module colors and labels for use in subsequent parts
#save(MEs, moduleLabels, moduleColors, geneTree, file = "../wgcna/networkConstruction-stepByStep.RData")
```


```{r}
# Define numbers of genes and samples
nGenes = ncol(datExpr);
nSamples = nrow(datExpr);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs0)

# Calculate correlation between eigengenes and treatment 
moduleTraitCor = cor(MEs, datTraits[,c("temp_num"), drop=F], use = "p");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

# Correlation among module eigengenes 
#cor(MEs0, use="complete.obs") %>% corrplot::corrplot(tl.cex=.75, diag = F)

# library(PerformanceAnalytics)
# chart.Correlation(MEs0[,], histogram=F, pch=19)
```

```{r}
sizeGrWindow(10,6)
# Will display correlations and their p-values
# textMatrix =  paste(signif(moduleTraitCor, 2), "\n(",
#                            signif(moduleTraitPvalue, 1), ")", sep = "");
textMatrix =  paste(signif(moduleTraitCor, 2), " (",
                           signif(moduleTraitPvalue, 2), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)

#pdf(file = "C:/Users/laura.spencer/Desktop/module-temperature-corr.pdf", width = 4.5, height = 6);
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(datTraits[,c("temp_num"), drop=F]),
               yLabels = names(MEs) %>% str_remove(pattern = "ME"),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.7,
               cex.main = 0.85,
               cex.lab.y = 0.75,
               cex.lab.x= 0.85,
               zlim = c(-1,1),
               main = paste("Module-temperature relationships"))
#dev.off()
```


```{r}
# Define variable containing the temperature column
temperature = as.data.frame(datTraits$temp_num);
names(temperature) = "temperature"

# names (colors) of the modules
modNames = substring(names(MEs), 3)

geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));

names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");

geneTraitSignificance = as.data.frame(cor(datExpr, temperature, use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

names(geneTraitSignificance) = paste("GS.", names(temperature), sep="");
names(GSPvalue) = paste("p.GS.", names(temperature), sep="");
```


```{r}
# Need to have every gene represented in the annotation file 
annot = counts.annot.pcod %>% 
  filter(gene %in% names(datExpr))
dim(annot)
names(annot)
probes = names(datExpr)
probes2annot = match(probes, annot$gene)

# The following is the number or probes without annotation:
sum(is.na(probes2annot))
# Should return 0.
```

```{r}
# Create the starting data frame
geneInfo0 = data.frame(id = probes,
                      SPID = annot$spid[probes2annot],
                      gene_pcod = annot$gene_pcod[probes2annot],
                      ncbi_id = annot$ncbi_id[probes2annot],
                      description = annot$description[probes2annot],
                      moduleColor = moduleColors,
                      geneTraitSignificance,
                      GSPvalue)
# Order modules by their significance for pH
#modOrder = order(-abs(cor(MEs, pH, use = "p")));
modOrder = order(-abs(cor(MEs, temperature, use = "p")));

# Add module membership information in the chosen order
for (mod in 1:ncol(geneModuleMembership))
{
  oldNames = names(geneInfo0)
  geneInfo0 = data.frame(geneInfo0, geneModuleMembership[, modOrder[mod]], 
                         MMPvalue[, modOrder[mod]]);
  names(geneInfo0) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep=""),
                       paste("p.MM.", modNames[modOrder[mod]], sep=""))
}
# Order the genes in the geneInfo variable first by module color, then by geneTraitSignificance
#geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.pH));
geneOrder = order(geneInfo0$moduleColor, -abs(geneInfo0$GS.temperature));

geneInfo = geneInfo0[geneOrder, ]
write.csv(geneInfo, file = "../wgcna/geneInfo.csv")
save(geneInfo, file="../wgcna/geneInfo")
load(file="../wgcna/geneInfo")
```

Visualizing the network of eigengenes

```{r}
# Recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes
# Add temp treatment to existing module eigengenes
MET = orderMEs(cbind(MEs, temperature))
# Plot the relationships among the eigengenes and pH

# Plot the dendrogram
sizeGrWindow(6,6);
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0),
plotHeatmaps = FALSE)

# Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
plotDendrograms = FALSE, xLabelsAngle = 90)


# # Barplots of eigengenes by sample for each pH-associated module
# # !!!! BROKEN - BARS AREN'T COLOR-CODED CORRECTION
# sizeGrWindow(8,7);
# which.module="red"
# ME=MEs[, paste("ME",which.module, sep="")]
# 
# all(rownames(datTraits)==rownames(MEs)) # check order of samples treatment and eigengene dfs are same. Yes.
# 
# par(mar=c(5, 6, 0, 0.7))
# barplot(ME, col=numbers2colors(as.numeric(datTraits$Treatment)), main="", cex.main=2,
# xlab="eigengene expression",ylab=NULL, names=rownames(MEs), horiz=T, las=1, cex.names=0.65)
# legend("topright", legend=rev(c(levels(datTraits$Treatment))), fill=rev(c(unique(numbers2colors(as.numeric(datTraits$Treatment))))))
# 

```

# departure from WGCNA tutorials 

## Generate line plots of genes for each module of interest
```{r}
pcod.blast
counts.annot.pcod
```


```{r}
# Generate per-gene summary stats 
# first extract the normalized counts
gene.counts.norm <- assay(vsd.temperature) %>% as.matrix() %>% as.data.frame() %>% rownames_to_column(var = "gene") %>% 
  pivot_longer(-gene, values_to = "counts.vsd", names_to = "sample_name") %>%
  left_join(sample.info.rna[,c("sample_name", "temperature")]) %>% 
  mutate_at(c("gene", "sample_name"), factor) %>%
#  mutate(temp_factor=factor(temperature, levels=c("0", "5", "9", "16"))) %>% 
  filter(sample_name %in% rownames(datExpr)) # <-- important, keep only samples that were retained in WGCNA analysis 

# calculate mean, sd, cv and annotate 
genes.stats <- gene.counts.norm %>% group_by(temperature, gene) %>% 
  summarize(mean=mean(counts.vsd), median=median(counts.vsd), sd=sd(counts.vsd), cv=sd(counts.vsd)/mean(counts.vsd)) %>% 
  mutate(DEG=if_else(gene %in% degs, "DEG", "Non-DEG")) %>%
  # mutate(change = case_when(
  # gene %in% c(rownames(subset(res.temperature.AM, padj < 0.05 & log2FoldChange < 0)), rownames(subset(res.temperature.AL, padj < 0.05 & log2FoldChange < 0))) ~ "Up-regulated",
  # gene %in% c(rownames(subset(res.temperature.AM, padj < 0.05 & log2FoldChange > 0)), rownames(subset(res.temperature.AL, padj < 0.05 & log2FoldChange > 0))) ~ "Down-regulated")) %>% 
  # mutate(change=replace_na(change,"Non-DEG")) %>% 
  mutate(DEG=factor(DEG, levels=c("Non-DEG", "DEG"))) %>% 
  # mutate(change=factor(change, levels=c("Non-DEG", "Down-regulated", "Up-regulated"))) %>%
  ungroup() %>% 
  left_join(counts.annot.pcod %>% dplyr::select(gene, gene_pcod, spid, gene_uni, ncbi_id, evalue, description), by="gene")
  #left_join(P.camt.blast.GO %>% dplyr::select(gene_id, spid, protein_names, evalue), by=c("gene"="gene_id"))

# Generate line plots of significant modules showing only genes significantly correlated to module 
# List of modules (their color IDs) that are significantly associated with pH 
#modules <- moduleTraitPvalue %>% as.data.frame() %>% filter(pH<0.05) %>% rownames() %>% str_remove(pattern = "ME")
modules <- cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr) %>%
  rownames() %>% str_remove(pattern = "ME")
modules.p <- paste("p.MM.", modules, sep = "")
modules.cor <- noquote(paste("MM.", modules, sep = ""))

save(modules, file="../wgcna/modules")
save(modules.p, file="../wgcna/modules.p")
save(modules.cor, file="../wgcna/modules.cor")

# create empty list to store summary statistics for each modules' sign. genes 
module.stats <- vector("list", length(modules))
names(module.stats) <- modules

# create empty list to store line plots for each modules' sign. genes 
module.lineplots <- vector("list", length(modules))
names(module.lineplots) <- modules

for (i in 1:length(modules)) {
a <- geneInfo %>%
  filter(moduleColor==modules[i]) %>% 
  filter(p.GS.temperature<0.05) %>% 
  filter(!!as.symbol(modules.p[i])<0.05) %>% 
  #filter(id %in% degs) %>%
  dplyr::select(id, moduleColor, GS.temperature, p.GS.temperature, !!as.symbol(modules.p[i]), !!as.symbol(modules.cor[i])) %>% 
  left_join(genes.stats, by = c("id"="gene"))
  #group_by(Treatment, moduleColor) %>% mutate(mean_z = scale(mean)) %>% ungroup() %>% #use this line of code to z-transform 

#formula <- y ~ x
#colors <- c("brown","pink","cyan2", "red","cyan4","green","darkgreen")
module.lineplots[[i]] <-  
  ggplot(a, aes(x=temperature, y=median)) + 
  geom_line(aes(group=id), size=.01, color=modules[i]) +
  theme_minimal(base_size = 8) + facet_wrap(~moduleColor) +  #change
  #geom_smooth(method="loess",se = FALSE, color="blue", size=1) + #  method=lm, formula = formula,
  #scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) + 
  theme(plot.title = element_text(size=8), axis.text.x = element_text(hjust = 0.2), 
        axis.title.x = element_text(size=6)) +
  xlab("temperature") #+
  # stat_poly_eq(aes(label = ..eq.label..),
  #              label.x.npc = 0.5, label.y.npc = 0.99,
  #              formula = formula, parse = TRUE, size = 3)

module.stats[[i]] <- a %>% dplyr::select(-temperature, -mean, -sd, -cv) %>% distinct()
}
#print(module.lineplots)
save(module.lineplots, file = "../wgcna/module.lineplots")

#Save module lineplots pdf
fig <- ggarrange(plotlist=module.lineplots,
                       ncol=2, nrow=3,
                       common.legend = TRUE, legend="top")
pdf(file = "../wgcna/module-lineplots.pdf", width = 8.5, height = 11, onefile = TRUE);
fig
dev.off()

modules.up <- c("darkgreen", "darkgrey", "orange", "bisque4", "darkred")
modules.down <- c("darkturquoise", "darkorange2", "darkmagenta", "greenyellow", "midnightblue", "thistle2", "salmon", "lightcyan")

module.lineplots
```

```{r}
geneInfo %>%
#  filter(moduleColor=="yellow") %>%
  filter(!!as.symbol(modules.p[i])<0.05) %>% 
  dplyr::select(id, moduleColor, GS.temperature, p.GS.temperature, !!as.symbol(modules.p[i]), !!as.symbol(modules.cor[i])) %>% 
  left_join(genes.stats, by = c("id"="gene"))
```

Summary of modules significantly associated with temperature

```{r}
moduleTraitCor %>% nrow()
modules %>% length()
cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr)
```
# For temperature-associated modules, look at correlation between module membership and gene significane for each gene 

```{r}
#par(mfrow = c(4,2));

for (i in 1:length(modules)) {
module = modules[i]
column = match(module, modNames);
moduleGenes = moduleColors==module;

#sizeGrWindow(7, 7);
verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]),
                   abs(geneTraitSignificance[moduleGenes, 1]),
                   xlab = paste("Module Membership in", module, "module"),
                   ylab = "Gene significance for Temperature",
                   main = paste("Module membership vs. gene significance\n"),
                   cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)
}

```
```{r}
ggplotly(geneInfo %>%
#  filter(moduleColor %in% modules) %>%
#  filter(p.MM.brown<0.05) %>%
  ggplot(aes(x=GS.temperature, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) +
  #scale_fill_manual(values=sort(modules)) +
  ggtitle("GS.temperature by module"))

ggplotly(geneInfo %>%
#  filter(moduleColor %in% modules) %>%
#  filter(p.MM.brown<0.05) %>%
  ggplot(aes(x=p.GS.temperature, group=moduleColor, fill=moduleColor)) + geom_density(alpha=0.5) +
  #scale_fill_manual(values=sort(modules)) +
  ggtitle("p.temperature by module"))
```

# Look at overlap among WGCNA and DESeq2 results 

## What percentage of DEGs are captured by pH-associated WGCNA modules? Particularly interested in genes I will use for enrichment analysis, so I will filter genes for: 
- Only genes whose abundances are significantly correlated with pH  (i.e. Gene Significance (GS) p-value, p.GS, <0.05)
- NO ---> Only genes whose module membership (MM) is significant (p.MM < 0.05). 


```{r, message=F}
# First how many DEGs do I have total? 
n.DEGs <- genes.stats %>% filter(DEG=="DEG") %>% dplyr::select(gene) %>% distinct() %>% nrow()

# merge module association info with DEG info 
geneInfo %>%
  left_join(genes.stats %>% dplyr::select(gene, DEG) %>% distinct(), by = c("id"="gene")) %>% #join with table indicating DEG
  mutate(moduleColor = as.factor(moduleColor)) %>%
  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.temperature<0.05), not.sign=sum(p.GS.temperature>=0.05)) %>% 
  filter(moduleColor %in% modules) %>% 
  filter(DEG=="DEG") %>% 
  ungroup() %>% dplyr::select(n, sign) %>% colSums()/n.DEGs*100 

# replace "modules" filter with  c(modules, "lightsteelblue") to include peak/valley module with lots of DEGs

# ANSWER: 
# 75.8% of the DEGs were assigned to a temperature-associated module. 
# 61.2% of the DEGs were both assigned to a temperature-associated module, AND those genes are sign. associated with temperature (p.GS.temperature<0.05)
```


```{r}
module.corr <- cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>% filter(pvalue<0.05) %>% arrange(corr) %>%
  rownames_to_column("module") %>% mutate(module=str_replace_all(module, "ME", ""))

# merge module association info with DEG info 
module.n <- geneInfo %>%
  left_join(genes.stats %>% dplyr::select(gene, DEG) %>% distinct(), by = c("id"="gene")) %>% #join with table indicating DEG
  mutate(moduleColor = as.factor(moduleColor)) %>%
  #group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.pH<0.05), not.sign=sum(p.GS.pH>0.05)) %>% 
  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.temperature<0.05), not.sign=sum(p.GS.temperature>=0.05)) %>% 
  filter(moduleColor %in% modules) %>% 
  left_join(module.corr, by=c("moduleColor"="module")) %>% ungroup()

# % of all DEGs that correlate positively with temperature (upregulated) at alpha < 0.025
module.n %>% filter(DEG=="DEG") %>% 
   filter(pvalue<0.025) %>% filter(corr>0) %>%
   dplyr::select(n, sign) %>% colSums()/n.DEGs*100 

# % of all DEGs that correlate negatively with temperature (downregulated) at alpha < 0.025
module.n %>% filter(DEG=="DEG") %>% 
   filter(pvalue<0.025) %>% filter(corr<0) %>%
   dplyr::select(n, sign) %>% colSums()/n.DEGs*100 
```

## What percentage of DEGs are also identified as correlated with temperature via WGCNA (and assigned to any module)

```{r, message=F}
geneInfo %>%
  left_join(genes.stats %>% dplyr::select(gene, DEG) %>% distinct(), by = c("id"="gene")) %>% #join with table indicating DEG
  mutate(moduleColor = as.factor(moduleColor)) %>%
  group_by(moduleColor, DEG) %>% summarise(n=n(), sign=sum(p.GS.temperature<0.05), not.sign=sum(p.GS.temperature>=0.05)) %>% 
  filter(DEG=="DEG") %>% 
  ungroup() %>% dplyr::select(sign, not.sign) %>% colSums()/n.DEGs*100 

# 71% of DEGs are deemed to be significantly correlated with pH
```

## How many genes are correlated with temperature (any module?)

```{r}
geneInfo %>%
  mutate(moduleColor = as.factor(moduleColor)) %>%
  group_by(moduleColor) %>% summarise(n=n(), sign=sum(p.GS.temperature<0.05), not.sign=sum(p.GS.temperature>=0.05)) %>% 
  dplyr::select(moduleColor, n, sign, not.sign) %>% 
  mutate(perc.sig=sign/n) %>% 
  dplyr::select(sign, not.sign) %>% colSums()
```

# Venn diagram of DEGs and genes in WGCNA pH-associated modules 

```{r}
display_venn <- function(x, ...){
  library(VennDiagram)
  grid.newpage()
  venn_object <- venn.diagram(x, filename = NULL, ...)
  grid.draw(venn_object)
}

module.genes <- vector("list", length(modules)+1)
names(module.genes) <- c(modules, "all")
for (i in 1:length(modules)) {
module.genes[[i]] <- geneInfo %>%
  # filter for genes assigned to temperature-associated modules 
  filter(moduleColor==modules[i]) %>% 
# further filter for genes that are significantly associated with temperature and significantly assigned to their respective module
filter(p.GS.temperature<0.05) %>%
filter(!!as.symbol(modules.p[i])<0.05) %>%
## alternative approach to just select "hub genes"   
  #    filter(abs(GS.temperature)>0.2) %>% filter(!!as.symbol(modules.cor[i])>0.8) %>% 
  rownames()
}
module.genes[[length(module.genes)]] <- Reduce(c, module.genes)

x <- list(
  Cold = row.names(diffex.0.filt), 
  Cool = row.names(diffex.5.filt), 
  Warm = row.names(diffex.16.filt),
  WGCNA = module.genes[["all"]]
  )

display_venn(
        x,
        category.names = c("DEGs\nCold" , "DEGs\nCool" , "DEGs\nWarm", "WGCNA\ntemperature-associated\nmodules"),
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#999999", "#E69F00", "#56B4E9", "#009E73"),
        print.mode="percent", sigdigs=2, #use this line to print percentages
        # Numbers
        cex = .9,
        fontface = "italic",
        # Set names
        cat.cex = .9,
        cat.fontface = "bold",
        cat.default.pos = "outer")#,
        #cat.dist = c(0.06, 0.06, 0.06))

# NOT UPDATED FOR THIS NOTEBOOK 
# EFFECT OF VARIOUS FILTERING SETTINGS: 
# When I include all genes assigned to pH-associated modules, the % of DEGs included in WGCNA modules is 80.1%
# When I filter for only genes that are significantly associated with pH (p.GS.pH<0.05), that drops to 73.3%
# When I filter for genEs that are sign. associated with pH (p.GS.pH<0.05) AND module membership is sign. (p.MM<0.05) that drops to 72.2%
# Because I am using all genes assigned to a pH-associated module in the GO Enrichment analysis and IPA, I shall use the least filtered set. 

```

Now look for overlap among DEGs and each module 

```{r}
par(mfrow = c(4,2));
for (i in 1:length(modules)) {
  y <- list(
  Cold = row.names(diffex.0.filt), 
  Cool = row.names(diffex.5.filt), 
  Warm = row.names(diffex.16.filt),
  WGCNA = module.genes[[i]]
  )
  
  display_venn(
        y,
        category.names = c("DEGs\nCold" , "DEGs\nCool" , "DEGs\nWarm", 
                           paste("WGCNA module: ", modules[i], "\ntemperature-associated\nmodules", sep="")),
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = c("#999999", "#E69F00", "#56B4E9", modules[i]),
        print.mode="percent", sigdigs=2, #use this line to print percentages
        # Numbers
        cex = .9,
        fontface = "italic",
        # Set names
        cat.cex = .9,
        cat.fontface = "bold",
        cat.default.pos = "outer")
}
```

# How many genes assigned to each temperature-associated module?

```{r} 
# How many total genes assigned to any temperature-associated module? 10,513 
geneInfo %>%  
  filter(moduleColor %in% modules) %>%
  dplyr::count(moduleColor) %>% dplyr::select(n) %>% colSums()

# How many total genes assigned to any temperature-associated module AND have high Gene Significance (significantly associated with temperature)?  7,437 
geneInfo %>%  
  filter(moduleColor %in% modules) %>%
  filter(p.GS.temperature<0.05) %>%
  dplyr::count(moduleColor)  %>% dplyr::select(n) %>% colSums()

```


# Generate heatmaps of temperature-associated modules

```{r}
# create empty list to store line plots for each modules' sign. genes 
counts.modules <- vector("list", length(modules))
names(counts.modules) <- modules

#filter(abs(GS.pH)>0.2) %>% filter(!!as.symbol(modules.cor[i])>0.8)

for (i in 1:length(modules)) {
  counts.modules[[i]] <- 
    (assay(vsd.temperature))[(geneInfo %>% 
                       filter(moduleColor==modules[i]) %>% 
                       filter(p.GS.temperature<0.05))$id,] %>% 
    as.data.frame() %>% add_column(module=modules[i])
}

s <- sample.info.rna %>% column_to_rownames("sample_name")

# define order of modules from most sign. positive correlation <--> most sign. negative correlation 
modules.sorted <- moduleTraitCor %>% as.data.frame() %>% 
  filter(rownames(.) %in% 
           rownames(moduleTraitPvalue %>% as.data.frame() %>% filter(temp_num<0.05))) %>% 
  arrange(temp_num) %>% rownames() %>% str_remove(pattern = "ME")

counts.modules.matrix<- bind_rows(counts.modules[modules.sorted]) %>% 
  dplyr::select(module, rownames(s)) %>%
  mutate(module=factor(module, levels=modules.sorted))

#pdf(file = "C:/Users/laura.spencer/Desktop/wgcna-modules-heatmap.pdf", width = 10, height = 7.5);
pheatmap(counts.modules.matrix[,colnames(counts.modules.matrix) !="module"], 
         cluster_cols = FALSE, cluster_rows=F, annotation_names_col = FALSE, 
         show_rownames=FALSE, na.rm=TRUE, scale="row", 
         main = "Expression by pH-associated module\n(significant genes only)", fontsize = 8, 
         # annotation_colors = list(Treatment=c("0"="#2c7bb6", "5"="#fdae61", "16"="#d7191c"),
         #                           module=c(magenta="magenta", lightcyan="lightcyan", firebrick4="firebrick4",darkviolet="darkviolet",royalblue="royalblue",
         #                                    coral1="coral1", plum1="plum1",pink="pink",blue2="blue2", lightsteelblue1="lightsteelblue1",ivory="wheat",
         #                                    lightgreen="lightgreen",purple="purple", green="green")),
         annotation_col=s[,"temperature", drop=F],
         annotation_row = counts.modules.matrix[,"module", drop=FALSE])
#dev.off()
```

What about the modules - that aren't sign. associated with temperature? any patterns there? 

Let's just look at ALL modules: 

```{r}
# Generate line plots of ALL modules showing only genes significantly correlated to module 

modules.all <- moduleTraitPvalue %>% as.data.frame() %>% rownames() %>% str_remove(pattern = "ME")
modules.all.p <- paste("p.MM.", modules.all, sep = "")
modules.all.cor <- noquote(paste("MM.", modules.all, sep = ""))

# create empty list to store summary statistics for each modules' sign. genes 
modules.all.stats <- vector("list", length(modules.all))
names(modules.all.stats) <- modules.all

# create empty list to store line plots for each modules' sign. genes 
modules.all.lineplots <- vector("list", length(modules.all))
names(modules.all.lineplots) <- modules.all

for (i in 1:length(modules.all)) {

a <- geneInfo %>%
  filter(moduleColor==modules.all[i]) %>% 
  filter(!!as.symbol(modules.all.p[i])<0.05) %>% 
  dplyr::select(id, moduleColor, GS.temperature, p.GS.temperature, !!as.symbol(modules.all.p[i]), !!as.symbol(modules.all.cor[i])) %>% 
  left_join(genes.stats, by = c("id"="gene")) %>% 
  mutate(temperature=as.numeric(as.character(temperature)))
  #group_by(Treatment, moduleColor) %>% mutate(mean_z = scale(mean)) %>% ungroup() %>% #use this line of code to z-transform 
  #filter(moduleColor %in% modules)

#formula <- y ~ x

modules.all.lineplots[[i]] <-  
  ggplot(a, aes(x=temperature, y=median)) + 
  geom_line(aes(group=id), size=.01, color=modules.all[i]) +
  theme_minimal() + facet_wrap(~moduleColor) + #change
  #geom_smooth(method="loess", se = FALSE, color="blue", size=1) + #method=lm, formula = formula, 
  #scale_x_continuous(breaks=c(7.6, 7.8, 8.0)) + 
  theme(plot.title = element_text(size=10), axis.text.x = element_text(hjust = 0.2)) +
  xlab("Temperature") #+
  # stat_poly_eq(aes(label = ..eq.label..), 
  #              label.x.npc = 0.5, label.y.npc = 0.99,
  #              formula = formula, parse = TRUE, size = 3)

modules.all.stats[[i]] <- a %>% dplyr::select(-temperature, -mean, -sd, -cv) %>% distinct()
}
print(modules.all.lineplots)

#modules.all.lineplots[["violet"]]
save(modules.all.lineplots, file = "../wgcna/modules.all.lineplots")
```

## Prep for DAVID Enrichment Analysis

```{r}
# Do enrichment on all modules, can filter out the ones that aren't correlated with temperature later 
# modules # modules with genes that respond (linearly?) to temperature 
# modules.all
# modules.all.cor
# modules.all[modules.all %!in% modules]

geneInfo.wgcna.degs <- vector("list", length(modules.all)) 
names(geneInfo.wgcna.degs) <- modules.all

geneInfo.david <- vector("list", length(modules.all)) 
names(geneInfo.david) <- modules.all
WGCNA.4David <- vector("list", length(modules.all))
names(WGCNA.4David) <- modules.all

for (i in 1:length(modules.all)) {
  geneInfo.wgcna.degs[[i]] <- geneInfo %>%   
#  geneInfo.david[[i]] <- geneInfo %>%   
    filter(moduleColor == modules.all[[i]]) %>% #keep all genes assigned to each module 
#    filter(moduleColor == modules.all[[i]] & p.GS.temperature < 0.05) %>% #only keep genes assigned to each module if they also correlate with temperature themselves
    filter(moduleColor == modules.all[[i]]) %>% filter(id %in% degs) %>% #only keep genes assigned to each module if they also are DEGs from DESeq2 (in any contrast)
  dplyr::select(id, SPID, gene_pcod, ncbi_id, description, moduleColor, GS.temperature, p.GS.temperature, !!as.symbol(modules.all.cor[i])) %>%
    remove_rownames() %>% 
#    filter(id %in% degs) %>% # optoinal, only include genes that are also DEGs 
#    arrange(desc(!!as.symbol(modules.cor[i])))
    arrange(p.GS.temperature)
  
  WGCNA.4David[[i]] <- geneInfo.david[[i]] %>% filter(!is.na(SPID)) %>%
    dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector()
}

# Write tab-delimited file with each column containing Uniprot IDs for each WGCNA gene set 
write_delim(x = plyr::ldply(WGCNA.4David, rbind) %>% t() %>% as.data.frame(), delim = "\t", col_names = F, na="", 
            file = "../wgcna/wgcna-4David.txt")

# background for DAVID - copy to clipboard to paste into DAVID 
geneInfo %>% dplyr::select(SPID, ncbi_id) %>% remove_rownames() %>% distinct() %>%
  dplyr::select(SPID) %>% na.omit() %>% unlist() %>% as.vector() %>% write_clip()


# In which module is the ZP3 gene?  greenyellow
geneInfo %>% filter(ncbi_id=="132464016") 

# In which module is the Tyoe-4 ice-structuring protein (antifreeze protein)?
geneInfo %>% filter(ncbi_id=="132451019")

# # Other code, could be used in future 
# for (i in 1:length(modules)) {
#   geneInfo.david[[i]] <- geneInfo %>%
#     mutate(sign = case_when((moduleColor == modules[[i]] & p.GS.temperature < 0.05) ~ !!as.symbol(modules.cor[i]), 
#       TRUE ~ 0)) %>%
#     filter(sign!=0) %>% 
#     dplyr::select(ncbi_id,SPID,sign) %>% remove_rownames() %>% 
#     dplyr::rename(!! sym(modules[i]) := 1) %>%
#     dplyr::rename(!! paste(sym(modules[i]), "sign", sep=".") := 2)
# }
```

```{bash}
ls ../wgcna/DAVID*.txt > ../wgcna/GO_filenames.txt
```

Prep a dataframe for joining with DAVID info

```{r}
module.genes.all <- vector("list", length(modules.all))
names(module.genes.all) <- modules.all
for (i in 1:length(modules.all)) {
module.genes.all[[i]] <- geneInfo %>%
  # filter for genes assigned to each module 
  filter(moduleColor==modules.all[i]) %>% 
  rownames()
}

module.info <- cbind(moduleTraitCor %>% as.data.frame(), moduleTraitPvalue %>% as.data.frame()) %>% dplyr::rename(corr=1, pvalue=2) %>%  
  arrange(corr) %>% rownames_to_column("module") %>% mutate(module=str_replace_all(module, "ME", "")) %>%
  left_join(sapply(module.genes.all, length) %>% as.data.frame() %>% setNames("n") %>% rownames_to_column("module"))
```


```{r, message=F, warning=F}
#### Read in enriched biological functions for all contrasts

filenames <- read_delim(file="../wgcna/GO_filenames.txt", delim = "\t", col_names = c("filename")) %>%
  mutate(module=gsub("../wgcna/DAVID_|.txt", "", filename))
file_list <- vector(mode = "list", length = nrow(filenames))
names(file_list) <- c(filenames$module)

for (i in 1:nrow(filenames)) {
    file_list[[i]] <- data.frame(read_delim(file=paste("../wgcna/", filenames[i,"filename"], sep=""), delim = "\t"))}

david.wgcna <- bind_rows(file_list, .id = "module") %>% clean_names() %>%  # Bind dataframe for each gene_set together
    #filter(category=="GOTERM_BP_DIRECT") %>%
    filter(p_value<0.05) %>%
    dplyr::rename(percent=x) %>%
    separate(term, into = c("term", "process"), sep = "~") %>%
    left_join(module.corr) %>%
    mutate(response=case_when(corr<0 ~ "Decrease", corr>0 ~ 'Increase', TRUE ~ "Not_Correlated")) %>%
    mutate(response=as.factor(response)) %>% 
  tidyr::complete(module, response, category) # add rows for gene_sets that are missing from dataframe

save(david.wgcna, file="../wgcna/david.wgcna")
```

Prep some dataframes for plotting 
```{r}
module.name.corr <- vector("list", length(modules.all))
names(module.name.corr) <- modules.all
for (i in 1:length(modules.all)) {
module.name.corr[[i]] <- paste(module.info[i,1],
                               " (n=", prettyNum(module.info[i,4], big.mark = ","), ")",
  "\ncorr= ", round(module.info[i,2], digits = 2), 
  ", pval= ", signif(module.info[i,3], digits = 2), 
                               sep="")}
```


```{r}
response <- c("Decrease", "Increase", "Not_Correlated")
response.mods <- vector("list", 3)
response.mods[[1]] <- (module.info %>% filter(corr>0 & pvalue<0.05))$module
response.mods[[2]] <- (module.info %>% filter(corr<0 & pvalue<0.05))$module
response.mods[[3]] <- (module.info %>% filter(pvalue>=0.05))$module

for (i in 1:3){
print(
  david.wgcna %>% View()
  #filter(module %in% modules) %>%
#  filter(count>=3) %>% 
#  filter(fdr<0.1) %>%
  filter(p_value<0.05) %>%
#  filter(p_value<0.01) %>%
#  filter(category=="GOTERM_BP_DIRECT") %>% 
  filter(category=="UP_KW_BIOLOGICAL_PROCESS") %>%
#  filter(category=="KEGG_PATHWAY") %>%
  group_by(module, response, genes, count) %>%
  dplyr::summarise(p_value=min(p_value)) %>% distinct() %>% ungroup() %>% #for rows with duplicate grouping variabes, select one with lowest p-value
  left_join(david.wgcna %>% dplyr::select(module, response, category, term, 
                            process, count, percent, p_value, fold_enrichment, fdr, genes), 
            by = c("module", "response", "genes", "p_value", "count")) %>%  #re-add data
#  ungroup() %>% arrange(module, response, p_value) %>% group_by(module, response) %>% dplyr::slice(1:15) %>%
  ungroup() %>% arrange(rev(module), p_value) %>%
  mutate(process = factor(process, unique(process), ordered = TRUE)) %>%
#  mutate(process = factor(process, rev(unique(process)), ordered = TRUE)) %>%
#  mutate(term=gsub(".*:", "", term)) %>% mutate(term=factor(term, rev(unique(term)), ordered=TRUE)) %>% 
  mutate(module=factor(module, ordered=T, levels=module.info$module)) %>%
#  mutate(module=factor(module, ordered=T, levels=rev(module.info$module))) %>%
  filter(response==!!response[i]) %>% 
  filter(process %!in% c("DNA integration", "DNA recombination")) %>%

ggplot(aes(y = process, x=module, col=module)) +  #for GO and Uniprot KW
#ggplot(aes(y = term, x=module, col=module)) + #for KEGG pathway 
  geom_point(alpha=0.65, aes(size=-log10(p_value))) + #size=count,  alpha=p_value, 
 facet_wrap(~category,scales="free", nrow = 2) +
   scale_color_manual(name="Module", 
                    values=c(greenyellow=darken("greenyellow", amount=0.25),
                            darkmagenta="darkmagenta",
                            darkorange="darkorange",
                            lightcyan=darken("lightcyan", amount = .35),
                            salmon="salmon",
                            thistle2=darken("thistle2", amount=0.35),
                            midnightblue="midnightblue",
                            darkturquoise="darkturquoise",
                            darkred="darkred",
                            orange="orange",
                            darkgreen="darkgreen",
                            bisque4=darken("bisque4", amount=.35),
                            darkgrey=darken("darkgrey", amount=0.2),
                            red="red",
                            floralwhite=darken("floralwhite", amount=.4),
                            brown="brown",
                            darkolivegreen="darkolivegreen",
                            skyblue3="skyblue3",
                            paleturquoise="paleturquoise",
                            sienna3="sienna3"),
                    #module.corr$module), #use all temperature-associated modules 
                    guide = guide_legend(override.aes = list(size=4))) +
  scale_size("-Log10 P-value", range = c(2,8), #breaks = c(2, 5, 10),
             guide = guide_legend(override.aes = list(col="gray50"))) +
  scale_x_discrete(drop=T) + #Do drop empty factors for temperature contrasts
  theme_cleveland() + 
  theme(
        legend.position = "right", 
        axis.text.x=element_blank(),
        axis.title.x=element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y=element_text(size=7.25), 
        plot.title = element_text(size=8),
        legend.text=element_text(size=6), 
        legend.title = element_text(size=6),
        strip.background = element_blank(),
        strip.text.x = element_blank()) +
  ggtitle(paste("Modules with genes that ", response[i], " with temperature", "\nBiological Process Uniprot Keywords", sep="")) +
#  ggtitle(paste("Modules that ", response[i], "\nBiological Process GO Terms", sep="")) +
  #  ggtitle(paste("Modules that ", response[i], "\nKEGG Pathways", sep="")) +
  guides(colour = guide_legend(override.aes = list(size=4, linetype="blank"), order=1)))
}


```


## Extract module eigengenes (first principal componenet) to relate to temperature treatment 

```{r, warning=F, message=F, error=F, }
#green$funct %>% unique()

# module.functions <- list(
#   'magenta'="(a) Protein production",
#   'lightcyan'="(b) Energy production",
#   'firebrick4'="(c) Steroid metabolism", #Carbohydrate metabolism,\nvesicle transport
#   'darkviolet'="\n(d) Tissue development",
#   'royalblue'="(e) Muscle development\nand activity",
#   'coral1'="\n(f) Immune function",
#   'plum1'="(g) Protein folding\n(stress response)",
#   'pink'="(h) Lipid & carbohydrate\nmetabolism, immune function",
#   'blue2'="(i) No enriched functions",
#   'lightsteelblue1'="(j) Transcription regulation",
#   'ivory'="(k) Signaling",
#   'lightgreen'="(l) Transcription regulation",
#   'purple'="\n(m) Transcription regulation",
#   'green'="(n) Transcription regulation,\nsignal transduction"
# )

#pdf(file="wgcna/Enriched-BP_WGCNA.pdf", height = 8, width = 4.5)
#tiff(file="../wgcna/Enriched-BP_WGCNA.tiff", height = 800, width = 425, res = 100)
eigengene.anno <- MEs %>% 
  rownames_to_column("sample_name") %>% 
  pivot_longer(cols = -sample_name, names_to = "module") %>%
  mutate(module=as.factor(gsub("^ME", "", module))) %>%
  #filter(module %in% modules) %>% #filter for only modules associated with temperature
  left_join(sample.info.rna %>% dplyr::select(sample_name, temperature)) %>%
  group_by(temperature, module) %>%
  summarise(mean=mean(value), sd=sd(value)) %>%
  mutate(temperature=as.numeric(as.character(temperature))) %>% 
# Add module temperature correlation info
  left_join(module.info) %>%
   mutate(change=case_when(
     (corr<0 & pvalue<0.05) ~ "Decrease",
     (corr>0 & pvalue<0.05) ~ "Increase",
     TRUE ~ "Not_Correlated") %>% as.factor()) %>%
  ungroup()
  

  #plot   
eigengene.anno %>%
  mutate(module.labels=factor(module, levels=module.info$module, labels=module.name.corr)) %>% # choose how to order module plots
  filter(change=="Increase") %>%  #choose which modules to plot 
  ggplot(aes(x=temperature, y=mean, colour=module)) +
  geom_errorbar(aes(ymin=mean-sd, ymax=mean+sd), width=.1) +
  geom_line(cex=.65) + geom_point() + 
#  scale_color_manual(values=c(Decrease="slateblue3", Increase="springgreen4", Not_Correlated="gray35")) + 
     scale_color_manual(name="Module", 
                    values=c(greenyellow=darken("greenyellow", amount=.25),
                            darkmagenta="darkmagenta",
                            darkorange="darkorange",
                            lightcyan=darken("lightcyan", amount = .35),
                            salmon="salmon",
                            thistle2=darken("thistle2", amount=0.35),
                            midnightblue="midnightblue",
                            darkturquoise="darkturquoise",
                            darkred="darkred",
                            orange="orange",
                            darkgreen="darkgreen",
                            bisque4=darken("bisque4", amount=.35),
                            darkgrey=darken("darkgrey", amount=0.2),
                            red="red",
                            floralwhite=darken("floralwhite", amount=.2),
                            brown="brown",
                            darkolivegreen="darkolivegreen",
                            skyblue3="skyblue3",
                            paleturquoise="paleturquoise",
                            sienna3="sienna3"),
                    #module.corr$module), #use all temperature-associated modules 
                    guide = guide_legend(override.aes = list(size=4))) +
 scale_y_continuous(limits = c(-0.2, 0.2), breaks = c(-0.2, -0.1, 0.1, 0.2)) +
 scale_x_continuous(limits = c(0, 16), breaks = c(0,5,9,16)) +
  xlab(NULL) + ylab(NULL) +
  theme_minimal() + 
  theme(legend.position = "none", strip.text = element_text(size = 10),
        panel.spacing.y = unit(-.1, "lines"),
        axis.text.y = element_text(size=9),
        axis.text.x = element_text(size=11)) +
  facet_wrap( ~module.labels, ncol = 2) 
#dev.off()  

```

